\chapter{Supplementary Figures for Latent Variables Models}
\label{ch:lvm_supp_figures}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/beta_contours_v325}
  \caption{The analog of Figure \ref{fig:beta_contours_v650} in the case that $V
    = 325$. \label{fig:beta_contours_v325}}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/beta_contours_nmf_d20}
  \caption{The results of the NMF experiment with $D = 20$ Within each panel, we
    display the true value of $\sqrt{\beta_{v}}$ as black points, while linked
    orange points give associated posterior medians. Note that the axes are
    truncated, and for some panels, the posterior medians all lie outside the
    visible box. Across columns, we vary and $\Earg{N}$. Along rows, we vary the
    assumed model, the inference procedure, and the true $p_{0}$ -- these are
    the three columns of row labels, read from outside
    in.\label{fig:beta_contours_nmf_d20}}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/beta_contours_nmf_d100}
  \caption{The analog of Figure \ref{fig:beta_contours_nmf_d20} when $D = 100$.
  \label{fig:beta_contours_nmf_d100}}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/beta_errors_nmf}
  \caption{A version of Figure \ref{fig:beta_errors_lda} for the NMF simulation
    experiment. The figures are read similarly, except there are a larger number
    of experimental configurations -- rows now distinguish the assumed model and
    shapes represent the true value of $p_{0}$. Further, while the first row of
    column labels still gives $D$, the second row gives $\Earg{N}$ instead of
    $N$. Note that we also now truncate the $x$ and $y$ axes, and not all points
    are visible. For example, most MCMC samples in the last panel in the second
    row have errors and SDs larger than what is displayed, and so are
    missing. \label{fig:beta_errors_nmf}}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/mu_intervals}
  \caption{
    A comparison of the posterior $p\left(\mu_{tv} \vert x\right)$ to the known
    underlying $\mu_{tv}$, in the unigram simulation experiment. The $x$-axis
    for each interval corresponds to the true $\mu_{tv}$ for one species at one
    timepoint, while the vertical intervals cover the 25\% to 75\% quantiles of
    samples from the posterior $p\left(\mu_{tv} \vert x\right)$. Different
    panels distinguish between configurations of $D$, $N$, $V$, and posterior
    sampling schemes. Posteriors from MCMC sampling seem to correctly recover
    the true underlying $\mu_{tv}$, while discrepancies arise for both VB and
    the bootstrap.
    \label{fig:mu_intervals} }
\end{figure}

\begin{figure}[ht]
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/mu_errors_unigram}
  \caption{A simplification of Figure \ref{fig:mu_intervals}, displaying RMSE when
    using posterior medians to estimate simulation $\mu_{tv}$s ($x$-axis) and
    the standard deviations of posterior marginals ($y$-axis), across
    experimental configurations. Generally, MCMC sampled posteriors seem to be
    the most reliable, across simulation configurations.
    \label{fig:mu_errors_unigram}
  }
\end{figure}

\begin{figure}
  \centering\includegraphics[width=\textwidth]{figure/lvm/visualize_lda_beta-F}
  \caption{Each credible interval describes an approximate posterior for one
    $\beta_{vk}$. Coupled with Figure \ref{fig:antibiotics_lda_theta}, this
    guides the interpretation of which bacterial taxa are more or less prevalent
    during antibiotic treatments. Each row of panels corresponds to one of the
    four topics, the $x$-axis indexes species, sorted according to phylogenetic
    relatedness, and the $y$-axis give transformed values of the species probability
    under that topic. Only the 750 most abundant species are shown. Note the
    disappearance of otherwise abundant species within Topics 2, 4, and to some
    extent, 1.}
  \label{fig:antibiotics_lda_beta}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/antibiotics_unigram_mu-F}
  \caption{Each posterior credible interval refers to one $\mu_{vt}$. The rows
    are a subset of times $t$ around the first antibiotic time course. The first
    row corresponds to a timepoint from before the treatment, the middle two
    from during the antibiotics time course, and the bottom from after the time
    course was stopped. Otherwise, this display is read in the same way as
    Supplementary Figure \ref{fig:antibiotics_lda_beta}. This view provides one
    way of smoothing abundance time series, to see how different species respond
    to antibiotic treatment. \label{fig:antibiotics_unigram_theta} }
\end{figure}

\begin{figure}
  \centering
  \includegraphics[scale=0.2]{figure/lvm/posterior_check_quantiles-F}
  \caption{As a posterior check, we compare the observed with simulated
    data quantiles, using a qq-plot. To reduce overlap, we have introduced a
    uniform $\left[0, 0.2\right]$ jitter on both axes. Further, the points are
    semi-transparent -- this makes it easy to see that most quantiles map to 0,
    which is expected, considering the sparsity of the data. From this view, we
    see that the LDA model tends to underestimate the overall number of zeros in
    the data, while the Dynamic Unigram model matches the observed quantiles
    almost exactly. \label{fig:antibiotics_posterior_quantiles} }
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/posterior_check_evals-F}
  \caption{As a posterior predictive check, we compute eigenvalues of data
    simulated from the fitted LDA model. The clouds of points summarize the
    posterior predictive distribution, while the black circles represent
    observed data eigenvalues. Note that the $y$-axis are logged eigenvalues.
    Evidently, the four-topic model effectively creates a rank-four
    approximation of the original data.
    \label{fig:antibiotics_posterior_evals}}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/posterior_check_scores-loadings-F}
  \caption{The eigenvalues displayed in Figure
    \ref{fig:antibiotics_posterior_evals} correspond to PCA results computed on
    posterior predictive samples, which are aligned and overlaid here. The left
    pair of panels give scores for each species, while the right pair provide
    loadings for each timepoint. The individual posterior samples have been
    smoothed into contours, while the posterior medians are displayed as shaded
    text. The observed data PCA results, after alignment with posterior samples,
    are displayed as black text. \label{fig:antibiotics_posterior_pca} }
\end{figure}

\begin{figure}
  \centering\includegraphics[width=\textwidth]{figure/lvm/visualize_lda_theta_boxplot-D}
  \caption{The analog of Figure \ref{fig:antibiotics_lda_theta} for Subject
    D. \label{fig:antibiotics_lda_theta_D}}
\end{figure}

\begin{figure}
  \centering\includegraphics[width=\textwidth]{figure/lvm/visualize_lda_beta-D}
  \caption{The analog of Supplementary Figure \ref{fig:antibiotics_lda_beta} for
    Subject D.}
\end{figure}

\begin{figure}
  \centering\includegraphics[width=\textwidth]{figure/lvm/visualize_lda_theta_boxplot-E}
  \caption{The analog of Figure \ref{fig:antibiotics_lda_theta} for Subject E.}
\end{figure}

\begin{figure}
  \centering\includegraphics[width=\textwidth]{figure/lvm/visualize_lda_beta-E}
  \caption{The analog of Supplementary Figure \ref{fig:antibiotics_lda_beta} for
    Subject E. \label{fig:antibiotics_lda_beta_E}}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/species_prototypes_1}
  \caption{Rather than displaying all representative species together, as in
    Figure \ref{fig:topic_prototypes}, we can sort species according to how
    representative they are of an individual topic. Here, the 15 species most
    strongly associated with Topic 1 are given. The panels are to be read from
    left to right and from top to bottom, to go in decreasing value of
    association. Note that the $y$-axis is on a square root
    scale. \label{fig:species_prototypes_1} }
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/species_prototypes_2}
  \caption{The analog of Figure \ref{fig:species_prototypes_1} for Topic
    2.\label{fig:species_prototypes_2}. }
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/species_prototypes_3}
  \caption{The analog of Figure \ref{fig:species_prototypes_1} for Topic
    3. \label{fig:species_prototypes_3} }
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/species_prototypes_4}
  \caption{The analog of Figure \ref{fig:species_prototypes_1} for Topic
    4. \label{fig:species_prototypes_4} }
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=0.8\textwidth]{figure/lvm/uneven_taxa_ordered}
  \caption{We can search for entire taxonomic families that seem associated with
    individual topics. Here we calculate the average of the topic
    representativeness statistic $\beta_{kv} - \sum_{k^{\prime} \neq k}
    \beta_{k^{\prime} v}$ across all species $v$ within each
    Family. Only those families that are most associated with a topic are
    displayed here. The sizes of circles represents the number of species within
    the family, which can be used to gauge the variability of the
    estimate. Compare this view with Supplementary Figure
    \ref{fig:uneven_taxa_facet}. \label{fig:uneven_taxa_ordered} }
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/lvm/uneven_taxa_facet}
  \caption{All species within the families screened out from Figure
    \ref{fig:uneven_taxa_ordered} are displayed here. Species among the
    representatives displayed in Figure \ref{fig:topic_prototypes} are colored
    according to the topics of they are prototypical. Grey series still
    contribute to the average family-topic association measure, but were not
    among the 50 prototypes for each topic. This view suggests that the
    Rhodospillaceae, Alcaligenaceae and Parabacteroides may have a large
    fractions of representatives from Topics 3, 1, and 2, respectively. Note
    that as before, abundances are plotted on a square root
    scale. \label{fig:uneven_taxa_facet} }
\end{figure}
