\chapter{Interactive Visualization of Microbial Counts}
\label{ch:interactive_vis}

We introduce methods for visualization of data structured along trees,
especially hierarchically structured collections of time series. To this end, we
identify questions that often emerge when working with hierarchical data and
provide an R package to simplify their investigation. Our key contribution is
the adaptation of the visualization principles of focus-plus-context and linking
to the study of tree-structured data.

Our motivating application is to the analysis of bacterial time series, where an
evolutionary tree relating bacteria is available a priori. However, we have
identified common problem types where, if a tree is not directly available, it
can be constructed from data and then studied using our techniques. We perform
detailed case studies to describe the alternative use cases, interpretations,
and utility of the proposed visualization methods.


\section{Treelapse}\label{introduction}

We introduce methods for visualization of data structured along trees,
especially hierarchically structured collections of time series. We hope both to
characterize generically useful techniques for interactively visualizing
hierarchical data and to offer practical tools for implementing such displays.
To this end, we identify questions that often emerge when working with
hierarchical data and provide an R package to simplify their investigation
\citep{rlanguage}.

In particular, we adapt the visualization principles of focus-plus-context and
linking to the study of tree-structured data \citep{buja1996interactive,
  becker1987brushing}. Our motivating application is to the analysis of
bacterial time series, where an evolutionary tree relating bacteria is available
a priori. However, we have identified common problem types where, if a tree is
not directly available, it can be constructed from data and then studied using
our techniques.

We have implemented our visualizations in D3, but encapsulated in an R package,
called treelapse, to facilitate rapid turnover from data preparation and
modeling to interactive exploration, and vice versa. Our code is open-source,
and is linked in the supplementary materials. We hope this package encourages
data analysts to work at the border between data modeling and visualization, and
more generally empowers a wider audience to apply less widely known, but
powerful, visualization ideas.

In summary, our key contributions are,
\begin{itemize}
\item Proposals for visualizing hierarchically structured data, based on
  principles from the data visualization community.
\item The implementation of these proposals in a publicly available R package.
\item The illustration of the wide reach of hierarchical data visualization,
  through case studies in both scientific and societal contexts.
\end{itemize}

The paper is organized as follows. First, we describe our motivating application
to the microbiome and the associated generic analysis tasks. Next, we review the
underlying visualization principles behind our contributions. Then we then
connect these principles to analysis tasks we identified earlier, describing in
detail the visualization methods we have implemented in treelapse. We close with
several case studies using publicly available data across both microbiome and
non-microbiome related applications.

\subsection{Problem Motivation}\label{problem-motivation}

A microbiome is a community of bacteria living in given environments, for
example, ocean water or the human gut \citep{human2012structure, cho2012human}.
Progress in the field has been rapidly accelerated by the advent of genomic
technologies, which enable detailed quantification of bacterial ecological
structure and its influence in human and environmental health. Being concerned
with both bacterial community structure and human health, the field exists at
the border between ecology and medicine; consequently, papers in the area often
apply a blend of exploratory data analysis and formal statistical inference.

The two essential microbiome analysis problems that motivated our work are the
tree-structured differential abundance and differential dynamics problems. In
the differential abundance problem, we attempt to compare the abundances of
individual bacteria across experimental conditions -- for example, treatment vs.
control or healthy vs. diseased. This is the microbiome analog of differential
expression analysis in genomics \citep{anders2010differential}. We prepend the
description ``tree-structured'' because, in practice, researchers generate
interpretations about intermediate taxonomic orders -- it is more interesting to
discover novel behavior taxonomic levels between high-order phyla and low-level
species. Hence, we frame the tree-structured differential abundance problem as
the question of identifying the largest taxonomic subtree whose associated
bacteria are differentially abundant.

\begin{figure}
  \centering
  \includegraphics[width=0.4\textwidth]{figure/treelapse/tree_differential_abundance}
  \caption{A cartoon of the tree-structured differential abundance problem. The
    tree represents the phylogenetic relationships between bacterial species,
    which are located at leaves in the tree. The two groups of interest (e.g.,
    treatment and control) are shaded in two colors, blue and green. Ignoring
    the different colors, the overall width of a tree branch encodes the amount
    of an observed sample that belongs to a group descending from that branch.
    The relative widths of the blue and green components within a branch
    corresponds to the relative frequency of that evolutionary group when
    comparing samples from either group. Differential abundance problem is to
    find large branches in this tree whose descendants have very different
    abundances between the two groups -- these are large branches where the
    split between colors is not even.
    \label{fig:tree_differential_abundance} }
\end{figure}

In the tree-structured bacterial dynamics problem, the goal is to describe
changes in bacterial abundances in an environment over time. As in the
differential abundance problem, it is useful if these descriptions can be given
at the highest subtree at which the pattern appears. Specific questions of
interest often have an ecological flavor. For example, researchers are often
interested in understanding how bacterial populations respond to sudden or
gradual environmental changes or how species fill, drop out from, or compete for
environmental niches. Medically, these questions are important for illuminating
the impact of antibiotic time courses or diet changes, for example.

\begin{figure}
  \centering
  \includegraphics[width=0.4\textwidth]{figure/treelapse/tree_differential_dynamics}
  \caption{
    The differential dynamics analog of Figure
    \ref{fig:tree_differential_abundance}. Instead of comparing two groups
    across many branches, we compare the average trajectories of taxa abundances
    across many phylogenetic positions. We expect trajectories to be similar
    across phylogenetically similar species -- rapid shifts in trajectories over
    a small phylogenetic distances are of high interest. Generally, the problem
    is to characterize variation in trajectories across phylogenetic space.
    \label{fig:tree_differential_dynamics}}
\end{figure}

\subsection{Problem Abstraction}\label{problem-abstraction}

To unify the tree-structured differential abundance and bacterial dynamics
problems, we identify the data with a collection of random variables indexed by
nodes in a prespecified tree structure. In the differential abundance problem,
each random variable lives in $\mathbb{R}^{G}$ where $G$ is the number of
groups being compared. Each coordinate represents the abundance for that group,
and a node exhibits differential abundance when the coordinates are drawn from
different distributions. On the other hand, in the bacterial dynamics problem,
each random variable is a time series, living in $\mathbb{R}^{T}$.

In both of these applications, we constrain the values of parent nodes according
to the value of the children nodes: we define the value at each node to be
either the sum or average of all descendant tips. However, it is possible to
imagine situations where the internal nodes are drawn from their own
distribution, unconstrained by descendants. In general, analysis in this
abstraction focuses on describing the distribution of these random variables as
a function of their position across subsets of the tree. The essential
difficulty in these problems is high-dimensionality -- there are many tree
nodes, each holding a vector-valued random variable. Even simply navigating
across the tree and comparing coordinates in the observed variables is a
challenge; ideally we could construct a succinct representation of the essential
covariation across subtrees and coordinates.

This framework suggests other potential application areas, not all of which have
a priori known tree structures. For example, collections of spatially-indexed
time series are frequently encountered in practice -- consider energy
consumption, product sales, or high school dropout rates across regional
districts. This type of data has an implicit tree structure -- at the top level
are different states, while at the bottom are individual census tracts, say.
Analysis here revolves around the question of how variation across time series
is related to their geographic position.

Alternatively, if this type of hierarchical contextual information is not
directly available, a tree structure can be learned from the data. This could be
achieved by learning a hierarchical clustering on the original series. Further,
if contextual information is available, but it is not hierarchical, it is
possible to setup a supervised problem that uses context to predict features of
the time series. We can construct a tree by applying a tree-based classifier
\citep{breiman1984classification} or extracting a regression tree from a more
complex supervised model \citep{boz2002extracting,saito2002extracting}. Analysis
then focuses on how different partitions of the contextual, covariate space
relate to observed time series.

Finally, note that, while we have focused on time series valued nodes,
all of this discussion could be translated to studying high-dimensional
data via parallel coordinates \citep{inselberg1991parallel}. The usual parallel
coordinates challenges remain, mainly selecting scales for and ordering
across the different coordinates, but the linking and focus-plus-context
can still be employed this setting.

\subsection{Background Literature and Solution Principles}
\label{background-literature-and-solution-principles}

Now that we have specified the essential questions of interest, we survey some
ideas from the visualization literature that can be applied to answer them. As
the core difficulty is high-dimensionality, it should be no surprise that the
techniques we adapt come from the literature on high-dimensional data
visualization, which has enjoyed rapid progress in the last 25 years. Modern
research in this domain develops abstractions and taxonomies for guiding
visualization designs so that they most effectively communicates properties of
the data to their intended audience. A major push in this body of work explores
the potential for interactivity to improve many stages of the data analysis
process, from preliminary data preparation, to refinement and navigation across
views, to final sharing and annotation of results \citep{heer2012taxonomy}.
Further work has attempted to bridge the gap between statistical analysis and
data visualization methodology, both of which provide techniques for learning
from high-dimensional data \citep{de2003visual}.

The problem structures most relevant to our study are tree and temporal
structure, and the visualization community has various ways of reasoning about
these data, see \citep{graham2010survey, aigner2011visualization} for detailed
surveys. From this literature, our approach is most directly informed by the
focus-plus context and linking principles, which we briefly review here. The
focus-plus-context principle is that large collections of visual elements can be
studied at multiple scales, by simultaneously focusing" on a few elements of
interest and maintaining the ``context'' of the coarser-scale background. A
simple example of this idea is to include a search box that highlights matching
samples (focus) and mutes the rest (context). Two more sophisticated methods
anchored in this idea are timeboxes and Degree-of-Interest (DOI) trees; both are
central to the proposals in treelapse \citep{hochheiser2004dynamic,
  heer2004doitrees}. In timeboxes, a collection of time series are graphically
queried using interactive brushes. Series that pass through all of the
user-specified brushes are highlighted, and the rest are faded to the
background. Hence, time series meeting the constraints imposed by the brushes
are focused, while the remainder are de-emphasized, though they remain present
as context. This method can be interpreted programmatically as the visual analog
of a database query, or probabilistically as the conditional distribution for
the full series, given it passes through certain bounds.

In DOI trees, the viewer's attention is focused on a collection of high-interest
nodes, while the remaining lower-interest nodes are left on the fringes as
context. The implementation is modularized into two tasks -- the determination
of a DOI distribution over nodes in the tree and visual layout of a tree given
DOI assignments. The DOI distribution used in \citep{heer2004doitrees} places
maximal interest on the node that the user had most recently clicked, along with
all ancestors. The DOI for all other nodes is defined as the distance to the
closest maximal interest node. The layout step then trims low-interest subtrees
until the remaining nodes fit within a given screen size. By adjusting the
minimal DOI below which nodes are hidden, the user can transition between
node-specific and full-tree scales.

\begin{figure}
  \centering
  \includegraphics[width=0.4\textwidth]{figure/treelapse/Internet_map_1024}
  \caption{An example of the focus-plus-context principle in action. The main
    view gives a global view of a part of the internet, while the inset zooms
    into a small neighborhood. Context is maintained via the inlay's border
    links. Image from The Opte Project, distributed under a Creative Commons
    license. \label{fig:internet_map}}
\end{figure}

In linking, alternative representations of the same samples are placed
side-by-side in order to display covariation across views. A canonical
application is to linked scatterplot brushing \citep{becker1987brushing}. Here,
a scatterplot matrix gives the relationship between all pairs of variables.
~Points brushed in one scatterplot are then highlighted in all others. For
example, this helps the user determine whether an outlier in one dimension is an
outlier in others. Another instance of this idea links the results of
dimensionality reduction methods to displays of the raw data, as implemented by
XGobi and Cranvas, for example \citep{xie2013cranvas, swayne1998xgobi}. As in
timeboxes, linking can be interpreted as database queries or conditional
probabilities: given a subset of the series after conditioning on the values for
one set of features, what are the values for a second set
\citep{buja1996interactive}?

\begin{figure}
  \centering
  \includegraphics[width=0.4\textwidth]{figure/treelapse/Ggobi04brushlink}
  \caption{
    An example of linked scatterplot brushing, as implemented by the GGobi
    package \citep{voigt2002extended}. The panel in the $ij^{th}$ position gives
    the scatterplot of variables $i$ and $j$ against each other. Brushing a
    set of points in the second panel in the first row highlights the same
    points across all panels.
    \label{fig:ggobi_brushlink} }
\end{figure}

Finally, unrelated to established visualization principles, we note that our
work is deliberately grounded in the R software ecosystem. This connection is
made using the htmlwidgets package \citep{vaidyanathan2014htmlwidgets}. Not only
does linking R with D3 make these visualization methods more broadly accessible,
we hope to facilitate exchange between data modeling and interactive
visualization. Moreover, our tools are intentionally limited in scope --
designed to facilitate this dialog for a specific class of problems, rather than
providing a toolbox for generic types of visualization design. We believe that
this narrow context within a broad ecosystem strikes a balance between
problem-specificity and ease-of-use.

\subsection{Specific Proposals}\label{specific-proposals}

Our first proposed visualization technique is a minor modification of the DOI
tree. The standard DOI tree definition does not have any notion of data defined
at nodes, it is only used a device for navigating tree structures. A trivial
extension can encode scalar data at nodes: have the node radius reflect the
associated scalar value. To reinforce this effect, we can adjust the width of
the parent edge. When parent nodes have values equal to the sum of their
children, this creates the effect of values ``flowing'' from the root to leaves.
To help viewers make use of their domain knowledge, we have included a search
box that highlights paths to nodes with matching terms. Edges are ordered from
widest on the left to narrowest on the right. While this method can only
represent a single scalar-value per tree node, it suggests an approach to the
tree-structured differential abundance problem, which we call the DOI sankey.

In the DOI sankey, we split each edge in the DOI Tree across different groups.
For example, suppose we have the average counts for treatment and control groups
at each tree node. Every edge in the tree is split into two colors\footnote{We
  use the colorbrewer palette to facilitate readability
  \citep{brewer2003colorbrewer}}, with relative widths of the different colors
reflecting differences in sizes for the two groups.

This display is designed to facilitate investigation of the tree structured
differential abundance question. For example, for a single node and a single
group, first compute the average abundance at that node among all samples in
that group. This will give the width for that group's color on the edge leading
to the specified node. Differentially abundant subtrees then correspond to
subtrees where some colors occupy more space than others. That is, this
representation makes it easier to identify points where the ``flows'' for
different groups diverge -- the colors begin to separate. The DOI principle
assists navigation across the tree structure, allowing focus on individual flow
structures without losing broader tree context.

Our third display is directed at the bacterial dynamics question. Here, two
panels are arranged one over the other; one displays all time series, while the
other displays all tree nodes, with node sizes reflecting the value at that node
averaged across all time points. For this reason, we call the display, timebox
trees. In the time series panel, we have directly implemented the timeboxes
idea. We then link the panels: when a set of series is highlighted by the
timeboxes, the associated tree nodes are also highlighted. For example,
timeboxes can be used to focus on a set of series with specific shape --
increased abundance after an ecological shock, for example -- and identify along
what subtrees this pattern is present. To further focus on specific elements, a
pan-zoom scented widget is provided \citep{willett2007scented}. The widget is a
miniature version of the full time series panel, equipped with a single brush
whose extent specifies the limits in the main time series panel. As in the DOI
trees and sankeys, a search bar can be used to highlight those series of
interest a priori.

The final display currently implemented in the package is the natural converse
of the timebox trees display. Rather than defining visual queries in terms of
time series, it defines queries using nodes in the tree. For this reason, we
call the display treeboxes. Rather than focusing on the intersection of brushes,
as in timebox trees, we focus on the union of brushed over nodes. This allows us
to highlight series associated with nodes on distant subtrees. This display is
also suited for the bacterial dynamics problem. For example, by highlighting all
nodes at one taxonomic level in the tree, we can easily summarize the time
series pattern for all the taxa at that level. Alternatively, focusing on all
the children below a single node makes it possible to see how much correlation
and competition there is between taxonomically similar bacteria. As in the
timebox trees display, a search box and pan-zoom scented widget are provided.

In principle, it would be desirable to combine the timebox and treebox displays,
so that highlighted nodes and series could be determined through brushes on both
the tree and series. For example, it would be useful to highlight the samples
that lie in the intersection of all timeboxes and union of all treeboxes. This
could allow more complex queries than are currently available. However, while
conceptually appealing, the authors encountered obstacles in practical
implementation: brush and mouseover events are required to occupy the same space
in this combined view. Properly distinguishing these events can be challenging,
and a solution based on the introduction of a lag between mousedown and the
manipulation of a brush led to a deteriorated user experience\footnote{However,
  the code for this approach is available publicly, in a separate branch of
  treelapse:
  \url{https://github.com/krisrs1128/treelapse/tree/combined-brushes}.}.

To be practically useful, the resulting visualizations must respond fluidly to
user interaction. As the data increase in scale, this fluidity can deteriorate
for two reasons. First, rendering many SVG elements in a framework like D3 is
costly\citep{johnson2008scalability}. While it is possible to use alternatives
-- HTML5's Canvas, for example -- it is often more challenging to implement
complex interactive behavior through them. Second, some of the dynamic queries
require a search over a many elements. These limitations are most pronounced in
the timebox tree display, which must search through all timepoints among all
time series whenever the brush is moved. The first, but not the second, concern
applies to treeboxes, while neither applies to DOI trees. Nonetheless, we feel
comfortable recommending timebox trees for data on the order of 500 tree tips
and 50 timepoints We note the sizes of the data sets in each case study below,
which each render fluidly, with the possible exception of the California housing
data, where there is a noticeable lag in the timebox tree and treebox displays.
In problems of larger size, we recommend a preliminary filtering or aggregation
step across nodes or, if the time series is smooth, across neighboring
timepoints, in order to avoid these potential scaling difficulties.

\subsection{Case Studies}\label{case-studies}

We now delve into applications on real data. Our goals are to illustrate
potential workflows that incorporate treelapse, describe the formulation of
questions that can be naturally investigated with our methods, and provide
example interpretations on treelapse output. Our examples are also chosen to
reflect the range of problem domains to which the package can be applied --
though it was motivated by applications to the microbiome, it is not tied to it.
More importantly, we argue that the visualization principles reviewed above can
substantively improve the practice of data analysis in the class of problems to
which we have limited ourselves.

\begin{table}
\centering
\begin{tabular}{|l|l|l|}
  \hline
  Data            & Number of timepoints & Number of nodes \\
  \hline
  Antibiotics     & 56                 & 386             \\
  Preterm births  & 216                & 318             \\
  Housing prices  & 254                & 944             \\
  Bikesharing     & 24                 & 819             \\
  Global patterns & 500                & 51             \\
  \hline
\end{tabular}
\caption{Problem dimensions for each of the case studies. For problems with
  dimensions larger than that in the housing prices problem, we recommend an
  initial summarization or filtering step to prevent performance
  issues. \label{problem-scaling}}
\end{table}

\subsubsection{Bacterial Dynamics of Antibiotics Time
Courses}\label{bacterial-dynamics-of-antibiotics-time-courses}

\citet{dethlefsen2008pervasive} investigated the effect of antibiotics on
bacterial community composition from an ecological perspective. The study tracks
the microbiome of three patients across ten months, with two five-day antibiotic
time courses separated by 6 months. Discerning the variation in resilience
across bacteria is important, considering the role of bacteria in health and
not just disease.

We approach the data using the linked time and treebox views, after first
filtering low variance taxa and taking an \(\text{asinh}\) transformation. An
initial view, Figure \ref{fig:antibioticoverview}, reveals two dramatic drops in
the overall bacterial abundance time series during the antibiotics time courses.
Two more subtle effects are also suggested from this view,

\begin{itemize}
\item
  The second antibiotic treatment seems to have a more lasting effect,
  as the series take longer to return to their original values.
\item
  Some high level taxa appear relatively unaffected by the first
  antibiotic treatment. By more closely inspecting the display, we are
  able to identify these as members of the Bacteroidetes phylum, see
  Figure \ref{fig:antibioticbacteroidetes}.
\end{itemize}

\begin{figure}
  \centering
  \includegraphics[width=375px]{figure/treelapse/annotated_antibiotic_overview}
  \caption{Here we display the primary timebox tree view of the antibiotics data
    set from \citep{dethlefsen2008pervasive}, annotated with the main components of the
    visualization. The tree at the top is a taxonomic tree of all the bacteria
    contained within the sample, and it is visually linked to the time series at
    the bottom: each node in the tree corresponds to a path among the time series.
    The selection brush is used to focus attention on the time series that go
    through it -- these are highlighted in blue -- and other brushes can be added
    using a button not displayed here. The pan-zoom widget at the top right is
    used to update the scales of the main time series display so that only
    particular time windows and $y$-axis ranges are visible. This view is the
    basis for all the timebox tree and treebox displays that appear
    below. \label{fig:antibioticoverview}}
\end{figure}

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/antibiotic_bacteroidetes}

}

\caption{Introducing a second box into the timebox display identifies the
  Bacteroidetes as a taxon only mildly impacted by antibiotics. The layout is
  identical to Figure 1, except two small brushes are placed over the time
  series between 10 and 20 days, and now only those time series and
  corresponding nodes in the tree are highlighted in blue. Further, the user has
  hovered over the top blue node in the tree, revealing the taxonomic identity
  of these series. Hence, brushing the time series and linking with the tree can
  be used to discover and characterize notable variation within the
  data.\label{fig:antibioticbacteroidetes}}
\end{figure}

Next, using the scented widget, we focus on the window around the second
antibiotic treatment. We apply the treebox display to compare then behavior of
different families of Firmicutes, Lachnospiraceae and Ruminoccocus. We suspect
that these taxa are associated with the delayed recovery after the second time
course. To investigate this, we input these family names in the search box to
isolate their positions on the tree; then we apply brushes to highlight the
series that contribute to these higher-level families. The resulting view is
given by Figure \ref{fig:antibioticfirmicutes}

\begin{figure}
{\centering \includegraphics[width=375px]{figure/treelapse/antibiotic_firmicutes}

}

\caption{Zooming into the second antibiotic time course and highlighting the
  Lachnospiraceae and Ruminococcus, we see that the Ruminoccocus took more time
  to recover to pre-treatment levels. Here, the red lines and nodes are those
  that match the text search provided by the user in a search box just below the
  figure (not displayed here). Hovering the mouse over these lines provides
  their identities -- the top red line is Lachnospiraceae, and the bottom red
  line is Ruminoccocus. Note that the brush in the treebox display is located
  over the tree, rather than over the time series. In particular, the search box
  and interactive brushing can be combined to interrogate hypotheses of a priori
  interest. \label{fig:antibioticfirmicutes}}
\end{figure}

Alternatively, we can summarize each node by the average across its descendants
-- this brings attention to individual bacteria that may be underlying some of
the broader taxonomic patterns we have noted when studying the subtree sums. For
example, in Figure \ref{fig:ruminococcus}, we highlight all families below order
Ruminococcus, suggesting that the decrease due to antibiotics occurs uniformly
across almost all families. A point that was not evident in the earlier
sum-across-descendants view is that, after the second treatment of antibiotics,
a few of the Ruminoccocus families recover more rapidly than the rest, for
example the Unc095d3 (highlighted in red) are only briefly affected. In
contrast, most families seem to recover in unison after the first treatment.

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/ruminococcus}

}

\caption{By hovering over the Ruminoccocus branches, we see that there is a
  prolonged effect of the antibiotics time courses more or less uniformly across
  the lower taxonomic orders. The graphical elements are the same as before,
  except the user has searched for Ruminoccocus and species Unc095d3, which has
  the highest average abundance within this taxon. By displaying averages rather
  than sums, we see that the effect of antibiotics visible at higher taxonomic
  orders is not created by a single dominant species becoming less abundant, but
  rather the decline in populations across all descendant species. The same
  display applied to different data can yield different
  insights.}\label{fig:ruminococcus}
\end{figure}

Further, note that in this subtree averages view, the tree display has changed.
This is because, at each branching point, we place the node with larger average
value on the left. Figure \ref{fig:verrucomicrobiae} notes that the nodes at
the far left in the tree are associated with phylum Verrucomicrobiae,
corresponding to a large average abundance across time points. This phylum had
been previously obfuscated -- because there are not many leaves associated with
this phylum, the sum was small. Interestingly, the abundance of these bacteria
seems to \emph{increase} after the first antibiotics treatment. Be cautious,
however, that the average over only a few Verrucomicrobiae species will be a
more variable estimate than the averages over the more prevalent phyla.

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/verrucomicrobiae}

}

\caption{The subtree averages aggregation brings attention to the
  Verrucomicrobiae, which though only present as a few species, are each rather
  abundant. In particular, they seem to increase after the first antibiotic time
  course, which occurs between days 15 and 20. This view was generated by
  placing a brush over the branch on the far left, which has those nodes with
  the largest averages across all timepoints. The user's mouse is over the blue
  series, which brings up the associated taxonomic label. The determination of
  species whose abundances increase during antibiotics, which would require many
  hypothesis tests using a more standard approach, becomes quickly apparent
  via interactive visualization.}\label{fig:verrucomicrobiae}
\end{figure}

\subsubsection{Differential Bacterial Abundance and Preterm
Births}\label{differential-bacterial-abundance-and-preterm-births}

\citet{digiulio2015temporal}
tracked the abundance of bacteria in the vaginal microbiome during
pregnancy in an effort to study relationships between bacterial
community composition and preterm birth. Ideally, it would be possible
to develop clear bacterial signatures associated with preterm births.

Unlike the antibiotics study, we have measurements across more
individuals than we could reasonably inspect one at a time. While we
could average across all individuals, we will take our cue from
\citep{digiulio2015temporal}
place each sample into one of five Community State Types (CSTs),
identified via k-medoids. In that study, a linear model identified one
of these CSTs (CST 4) as significantly more diverse, further it appeared
associated with preterm births. Here, we corroborate this finding using
exploratory views.

Therefore, our focus here is on the differential abundance question,
rather than dynamics. We would like to provide visual representations of
differential abundance across CSTs and also between preterm and non-preterm
births. \citet{digiulio2015temporal} interpreted the CSTs using a heatmap, with
bacteria ordered according to a hierarchical clustering. By using the DOI sankey
instead, we can interpret the CSTs in their taxonomic context and at multiple
scales of taxonomic resolution. Further, while
\citet{digiulio2015temporal} focused on identifying associations between
preterm births and CSTs -- presumably because testing individual bacteria loses
power -- we can compare bacterial abundances between preterm and non-preterm
samples along subtrees, without requiring CSTs as an intermediary.

In Figure \ref{fig:pretermcsts}, we compare the 5 CSTs according to
their values along the subtree. Specifically, we took the average of all
samples within each CST to define values at the (species-level) leaf
nodes, and then aggregated the averages up to the root. It is
immediately clear that samples from CST 4 have much more taxonomic
diversity. Further, focusing on the Lactobacillaceae family, we note
that the differential abundance of these bacteria distinguishes the
remaining CSTs, see Figure \ref{fig:pretermcstslacto}.

Alternatively, in Figure \ref{fig:pretermpreterm}, we avoid working with CSTs,
displaying instead averages among samples associated with either preterm or term
births. The green and yellow edges are associated with preterm births -- we see
that they contribute more weight to phyla outside the Firmicutes. This is
consistent with the claim that CST 4, the most diverse of the CSTs, is
associated with preterm births.

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/preterm_csts}

}

\caption{The increased diversity among samples in community state type (CST) 4
  is represented by the relatively larger contribution of red edges to branches
  outside of the Firmicutes. This display shows the top of the DOI sankey
  visualization of the preterm birth data studied in
  \citep{digiulio2015temporal}. The root of the tree is the taxonomic kingdom
  Bacteria, its children are labeled according to their phylum names. Each color
  within a branch is associated with CST, and the width of the associated color
  corresponds to the average abundance of that taxonomic group among all samples
  belonging to the corresponding CST, as indicated by the legend at the left.
  Phyla are sorted from most to least abundant. This initial display of the DOI
  sankey provides a summary of overall abundances across taxonomic groups and
  CSTs, and suggests subtrees to navigate into, to extract more detailed
  abundance information.
}\label{fig:pretermcsts}
\end{figure}

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/preterm_csts_lacto}

}

\caption{Zooming into the Lactobacillaceae family, we notice that the difference
  between the remaining four CSTs is related to which types of Lactobacillus are
  most prominent. The DOI sankey refocuses the tree around the last group that
  was clicked on, in this case showing more detail about order Lactobacillales
  and its descendants. The overview display can be recovered by navigating back
  up to higher-level taxa. Hence, it is possible to navigate between broad
  overview and detailed displays in a way that facilitates interpretation of
  results from statistical analysis.}\label{fig:pretermcstslacto}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=300px]{figure/treelapse/preterm_preterm}
\caption{Samples with high levels of phyla other than Firmicutes appear to be
  related to preterm births. Here, we again display the preterm birth data with
  a DOI sankey, but instead of grouping samples according to statistically
  generated CSTs, we directly assign samples to preterm vs. not preterm
  according to whether the mother eventually had a preterm birth. These new
  states are visible in the updated legend. Through interactivity, it becomes
  possible to develop meaningful visual summaries even before calculating formal
  statistical ones.}
\label{fig:pretermpreterm}
\end{figure}

\subsubsection{Dynamics in Housing Prices}\label{zillow-study}

We next consider an application unrelated to the microbiome, but with relatively
clear hierarchical structure. Our data are downloaded from Zillow and give the
Zillow Home Value Indexes at the neighborhood level, across the country,
computed monthly between 1996 and 2016. A link to the data source is provided in
the supplementary materials. In our display, we have taken the natural log of
these indexes. As our hierarchical structure, we use each neighborhood's
assignment to state, regional, county, and city levels. We represent each of
these coarser spatial categories using the average of all neighborhoods
contained in them. We have filtered down to the 890 neighborhoods in California;
rendering more neighborhoods while keeping all 246 timepoints causes a severe
lag in the interface.

Our basic analysis revolves around geographic and temporal variation in home
prices. We are especially interested in the effect of the 2008 recession and any
variation in the lead-up to or recovery from this event. These questions can be
naturally framed using timebox trees and treeboxes.

For example, we can study the trajectories of home prices among neighborhoods,
conditional on their being middle-income before the recession. We generate the
sequence of views in Figures \ref{fig:zillowmiddlepre},
\ref{fig:zillowmiddleup}, and \ref{fig:zillowmiddledown} to this end. The first
of these figures isolates neighborhoods with middle incomes before the
recession, using a single timebox. Since there appears to be a divergence in
trajectories after the recession, we introduce a second post-recession timebox,
dragging it over series with higher and lower incomes during this second time
period. This is the content of Figures \ref{fig:zillowmiddleup} and
\ref{fig:zillowmiddledown}. Though not directly visible from the static figures,
hovering the mouse over the highlighted tree nodes provides the geographic
identities, and we find that most of the middle-income series that increased
after the recession are associated with middle-income neighborhoods within the
coastal Southern California counties. For example, the mouse is currently over a
subtree with many highlighted nodes, which is shown to be the Los Angeles-Long
Beach-Anaheim metropolitan area. In contrast, hovering over nodes associated
with those middle income neighborhoods that saw decreases indicates that they
were mostly located in Central California and Oakland. In Figure
\ref{fig:zillowmiddledown}, the mouse is positioned over the Sacramento (which
is located in Central California) subtree, and seems enriched for this subset of
strongly recession-affected series.

The previous analysis highlights the fact that, within even narrow geographic
regions, there can be substantial variation in prices. We can study this
directly using treeboxes. In Figure \ref{fig:zillowsf} we have highlighted all
series in San Francisco County. We see that, in 2016, prices range from around
$e^{13} \approx \$440,000$ to $e^{14.5} \approx 2 \text{ million}$. So,
while all these neighborhoods tend to be among the more expensive ones in
California, prices can vary in a non-smooth way across geographic space.

We conclude this example with a caveat that the Zillow data are not
representative of all neighborhoods in California, only those with enough
listings on the site, so should be supplemented by other data sources for any
substantial decision-making.

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/zillow_middle_pre}

}

\caption{The time series here represent California neighborhood home prices
  between 1996 and 2016, and the tree corresponds to a geographic hierarchy,
  with regions at the top and neighborhoods at the bottom. We have brushed the
  neighborhoods with mid-range home prices before the recession. The associated
  tree nodes are highlighted at the top. Note that the collection of series
  seems to widen after 2008 -- we are interested in whether there are reliable
  predictors of these alternate trajectories, given their similar starting
  points. This serves as a baseline with which to compare Figure
  \ref{fig:zillowmiddleup} -- these views are easy to transition between
  interactively.}\label{fig:zillowmiddlepre}
\end{figure}

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/zillow_middle_up}

}

\caption{Among those neighborhoods with mid-range prices before the recession,
  displayed in Figure \ref{fig:zillowmiddlepre}, we have selected those that
  recovered more rapidly by introducing a brush at the top right of the time
  series panel. By hovering a brush over a collection of tree nodes that all
  seem to be highlighted, we infer that the associated neighborhoods are located
  mainly in Los Angeles and San Diego counties. This follow-up view is
  interesting to contrast with Figure \ref{fig:zillowmiddledown}.
}\label{fig:zillowmiddleup}
\end{figure}

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/zillow_middle_down}

}

\caption{In contrast to Figure \ref{fig:zillowmiddleup}, we can follow-up the
  selection in Figure \ref{fig:zillowmiddlepre} by isolating those neighborhoods
  where prices remained depressed after the recession. This is accomplished by
  moving the brush on the right down towards lower prices. Hovering over the
  associated highlighted nodes in the tree to reveal the associated locations,
  we see that most of these series correspond to neighborhoods in Central
  California and the East San Francisco Bay Area. The ability to sketch the
  overall shapes of series using brushes and interpret the associated selections
  using a tree simplifies what might otherwise be complex
  comparisons.}\label{fig:zillowmiddledown}
\end{figure}

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/zillow_sf}

}

\caption{In contrast to Figures \ref{fig:zillowmiddlepre} to
  \ref{fig:zillowmiddledown}, we can study the variation in series associated
  with a subset of the tree by using treeboxes. To study the range in home
  prices within San Francisco County, we can brush over the associated nodes in
  the tree. The red circle indicates that the user has searched for ``San
  Francisco,'' which guides the user to the appropriate subtree. The mouse is
  hovering over one of the more expensive neighborhoods. Hence, though the
  timebox tree and treebox views are similar, they are directed towards
  different types of visual comparisons.}\label{fig:zillowsf}
\end{figure}

\subsubsection{Sources of Variation in Bikesharing
Demand}\label{bikesharing-study}

Our next example is a study in bikesharing demand, included as an example of
analyzing collections of time series when there is no obvious hierarchical
structure a priori. The data are available at the UCI Repository and were
originally collected by a Washington D.C.-based bikesharing system for use in a
Kaggle prediction competition. A link is provided in the supplementary
materials. The data are hourly measurements of bike demand, aggregated across
all bikesharing stations, over two years, along with supplemental weather data.
In the competition, participants were asked to predict the hourly demand on a
held-out test set. Here, we adopt a descriptive view instead, attempting to
characterize factors associated with variation in bikesharing demand.

Like the Zillow home prices application, we study this problem as one of
describing a large collection of related time series. Here, we consider
the demand during a single day to be one time series; this is a natural
choice considering the daily periodicity of bike demand. To arrange
these daily series along an interpretable tree structure, we apply a
regression tree relating the supplemental data to the bikesharing demand
\citep{breiman1984classification}. In more
detail, we built this tree by noting the ``two table'' structure of this
problem: one describes bike demand, the other holds the supplemental
data. In both, the rows index days, while the columns index either hours
or supplemental features. Our tree is the trained regression tree after
predicting demand at 8AM based on the supplemental data. We choose this
response because (1) we need a univariate response in order to apply
regression trees and (2) the more straightforwards reduction to
daily-average-demand fails to distinguish between weekdays and weekends,
whose series appear qualitatively very different from each other.

Given this response, the first split in the regression tree is
(unsurprisingly) the difference between weekends and weekdays. This is
emphasized in Figures \ref{fig:working} and \ref{fig:weekend}, respectively;
using timeboxes to isolate the two types of series highlight the left and right
sides of the tree, respectively. For a more subtle effect, we select the
internal nodes associated with the first split below the weekday vs. weekend
split; these are given in Figures \ref{fig:weekday2011} and
\ref{fig:weekday2012}. This suggests that weekday demand increased during the
second year.

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/working}

}

\caption{The two peaks at rush hour distinguish weekday series from the rest.
  through the timebox tree view. The display is the same type of timebox tree
  view introduced in Figure \ref{fig:antibioticoverview}, but applied to the
  bikesharing data, where the time axis represents the time of day and the
  $y$-axis gives bikesharing demand. Each series is the bikesharing demand for a
  single day, over the course of two years. The tree now corresponds to the
  regression tree generated by predicting demand at 8am using supplementary
  data. Two brushes are introduced to highlight the double peaks corresponding
  to rush hours on weekdays. We see that although hierarchical structure was not
  present immediately in the bikesharing data, it is useful to introduce and
  interpret such structure by combining regression and visualization
  methodology.
}\label{fig:working}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=375px]{figure/treelapse/weekend}
    \caption{By adjusting the two brushes in Figure \ref{fig:working}, we see that
      unlike weekday demand, weekend demand is unimodal. The few weekday series with
      unimodal series seem to be associated with holidays. This is the case for New
      Years' Eve, which is currently hovered over in the tree. The ease of
      transitioning from Figure \ref{fig:working} to this display indicates the
      importances of brushing in interaction.}\label{fig:weekend}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=375px]{figure/treelapse/weekday_2011}
    \caption{Weekday demand appears larger in 2012 than 2011 -- compare with Figure 
      \ref{fig:weekday2012}. Here, a brushes are introduced over the tree to see the
      series associated with a particular split point. The red nodes are the results
      of searches over the two nodes that are children of this split point. The fact
      that the \texttt{yr>=0.5} selected line is larger than the \texttt{yr<0.5}
      line means that demand was larger in 2012. In combination with searching and
      treeboxes, it is possible to interpret more subtle split points in the decision
      tree.} \label{fig:weekday2011}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=375px]{figure/treelapse/weekday_2012}
  \caption{Weekday demand increased in 2012 -- compare with 
    Figure \ref{fig:weekday2011}. The search terms are the same as in that figure,
    but the subtree associated with the 2012 split point is highlighted, using the
    union of two boxes. Note that unlike timebox trees, which highlighted series
    lying through the intersection of brushes, treeboxes highlight series within the
    union of brushes.}\label{fig:weekday2012}
\end{figure}

In contrast to these general questions about daily demand, we could ask
a more granular question about specific time windows. For example, what
characterizes days on which there is larger than average demand after
midnight? We can select these series after first zooming into this time window.
Figure \ref{fig:warmweekend} reveals that the highlighted series are associated
with the warm-weekend split, which seems quite reasonable in retrospect.

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/warm_weekend}

}

\caption{The samples with highest night demand tend to fall on warm
  weekends. Here, the pan-zoom widget has been used to adjust both time and
  demand axes to narrow on a specific window of interest. Brushes are drawn over
  the larger among these series, and the corresponding tree nodes are located
  close to one another, in the part of the tree corresponding to the warm / cold
  average temperature split. More generally, panning and zooming allows
  navigation between focus and context.}\label{fig:warmweekend}
\end{figure}

Finally, we can study the behavior of the regression tree itself using
the DOI sankey. Here, we group samples according to their quintile of
8AM demand and then count the abundance of the groups flowing down
different branches. We find that the quintiles are each rather strongly
separated after descending even a few steps down the regression tree --
for example, Figures \ref{fig:weekday2011} and \ref{fig:weekday2012}
focus on 2011 vs. 2012 split among weekday samples, showing that this
split distinguishes between samples falling in the second and third
quintiles of 8AM demand.

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/bike_sankey}

}

\caption{So far, we have focused on the timebox tree and treebox representations
  of the bikesharing data -- a complementary view is provided by the DOI sankey.
  Here, the tree is the result of the regression tree procedure, while the
  colors represent particular quantiles of 8am demand. This allows the
  determination of which split descendants are associated with low or high
  demand.}\label{fig:bikesankey}
\end{figure}

This interactive representation of regression trees is potentially more
useful on larger trees that cannot be easily parsed in a single view; in
this sense the bikesharing tree is relatively simple. In our ideal data
analysis workflow, we imagine the analyst applying interactive
visualization and modeling techniques in an iterative, nonlinear
fashion, in the spirit of \citep{de2003visual}.

\subsubsection{Hierarchically Clustering the Global Patterns Data}\label{global_patterns}

Each of the timebox tree and treebox examples presented so far have
focused on data with a clear time component. We note however that these
techniques could alternatively be applied to high-dimensional data, via
the use of parallel coordinates \citep{inselberg1991parallel}. The usual
parallel coordinates challenges remain, namely selecting scales for and an
ordering across the different coordinates, but the linking and
focus-plus-context ideas can still be employed in this setting. Here we
provide an implementation of this idea on a dataset comparing microbiomes across
various ecological environments \citep{caporaso2011global}, which is publicly
accessible through the phyloseq R package \citep{mcmurdie2013phyloseq}.

The original Global Patterns data consists of 26 samples across 9
environments (for example, freshwater and soil). In each site, there are
counts across 19216 taxa -- to simplify visualization, we filter to the
500 most abundant taxa.

We hierarchically cluster these 26 samples based on the 500 most
abundant taxa, using complete linkage on the UniFrac distance. Figure
\ref{fig:gptimebox} displays the resulting hierarchy together with a
parallel coordinates view of the \(\text{asinh}\) transformed taxa.

\begin{figure}
  {
    \centering
    \includegraphics[width=225px]{figure/treelapse/gp_cluster1}
    \includegraphics[width=225px]{figure/treelapse/gp_cluster2}
}

\caption{An application to the Global Patterns demonstrates how linking in
  treelapse can be applied to combine hierarchical clustering and parallel
  coordinates views. Each panel represents a different subtree cluster within
  this data set, as indicated by the different locations for the tree brushes.
  The paths in the lower halves of each display represent the average value
  across different bacteria rather than timepoints, as in all previous figures.
  Though the samples originally do not include any hierarchical structure,
  hierarchical clustering provides such a structure which can then be
  interpreted using treeboxes.}\label{fig:gptimebox}
\end{figure}

In Figure \ref{fig:gptimebox}, we compare two subclusters from the
hierarchical clustering tree, after zooming to a few of the bacteria
that distinguish between the clusters. In contrast to the figures displayed to
this point, we only print time series associated with observed samples -- the
leaves of the hierarchical clustering tree. This reduces visual artifacts that
can be created by plotting many similar internal nodes, and which can overwhelm
patterns occuring in the leaves, which are those of central interest. Upon
revisiting the original data, it becomes clear that the samples highlighted on
the left come from freshwater samples, while those on the right come from soil
and skin, and looking up taxonomic groups associated with the distinguishing
bacteria confirms this. For example, many of the species with high
abundances in the left figure come from order Oceanospirillales.

\subsubsection{Inspecting Confirmatory Analysis}\label{structssi}

In addition to facilitating exploratory study, treelapse has potential
value as a device for inspecting confirmatory analysis. We provide an
illustration extending an example from \citep{callahan2016bioconductor}, which
formally tested bacterial species for association with age in a sample of mice.
The testing approach advocated there is particularly well-suited to
visualization with treelapse, since it sought to detect associations at multiple
levels of phylogenetic resolution, using statistical tools developed in
\citep{yekutieli2008hierarchical, sankaran2014structssi}.

The data of interest in \citep{callahan2016bioconductor} are bacterial counts
collected across old and young mice. After variance-stabilizing these counts
using DESeq2 \citep{love2014moderated}, a $t$-test was applied to each node in a
phylogenetic tree, comparing abundances between old and young mice. To account
for multiple testing, we employ the structSSI algorithm
\citep{yekutieli2008hierarchical, sankaran2014structssi} along with methods
available in the \texttt{multtest} package \citep{pollard2005multiple}.

To interpret the results, we apply timebox trees. Our goals are to (1) identify
subtrees with consistently elevated differential abundance across age groups and
(2) compare alternative multiple testing adjustment procedures. Our approach is
to display the negative-log raw and adjusted $p$-values for each node, with
alternative methods compared via parallel coordinates. One view of the resulting
display is captured in Figure \ref{fig:structssi}. First, we see that
significant nodes tend to be significant across all methods -- the ordering
between different series appears stable. Interestingly, the Sidak one-step and
structSSI procedures seem to have lower power than the others, including
conservative FWER-controlling methods, like the Bonferroni procedure. Further,
in this application, FDR-controlling techniques do not seem to offer notably
different adjusted $p$-values, relative to those controlling FWER. This suggests
that, for this problem, bacteria are either strongly associated with age, or not
associated at all, so that there is little gain from using more sensitive
procedures.

\begin{figure}

{\centering \includegraphics[width=325px]{figure/treelapse/structssi}

}

\caption{Viewing a tree of $p$-values across different methods highlights two
  subtrees with strong associations with mouse age, across several testing
  procedures. The tree represents the taxonomy of bacteria, and the series
  provides the negative log $p$-values associated with nodes as computed by
  different tests, listed along the $x$-axis as in parallel coordinates. By
  selecting series with larger values for a test, we see the associated subtrees
  of significant $p$-values. Hence, hierarchical views can be useful even in the
  confirmatory testing settings which typically study results from individual
  tests in isolation from each other.\label{fig:structssi}}
\end{figure}

Further, selecting series with strong association between abundance and age, two
major subtrees are brought to the forefront. Separately querying the
taxonomic identities of these bacteria reveals that they are two subgroups of
Clostridia, which is consistent with the analysis of
\citep{callahan2016bioconductor}. More than this specific analysis outcome, this
view demonstrates that interactive visual inspection of results from
confirmatory analysis provides deeper insight than the standard practice of
printing tables of (adjusted or unadjusted) $p$-values: the relationship between
significant nodes is only clear upon visualization on the tree.

\section{mvarVis}

mvarVis is an R package for visualization of diverse multivariate analysis
methods. We implement two new tools to facilitate analyses that are cumbersome
with existing software. The first uses htmlwidgets and D3 to create interactive
ordination plots, while the second makes it easy to bootstrap multivariate
methods and align the resulting scores. These interactive visualizations offer an
alternative to printing multiple plots with different supplementary information
overlaid, and bootstrapping enables qualitative assessment of the uncertainty
underlying the application of exploratory multivariate methods on particular
data sets.

Our approach is to leverage existing packages -- FactoMineR, ade4, and vegan
\citep{le2008factominer, dray2007ade4, oksanen2007vegan} -- to perform the
actual dimension reduction, and build a new layer for visualizing and
bootstrapping their results. This allows our tools to wrap a variety of existing
methods, including one table, multitable, and distance-based approaches --
principal components, multiple factor analysis, and multidimensional scaling,
for example. Since our package uses htmlwidgets, it is possible to embed our
interactive plots in Rmarkdown pages and Shiny apps
\citep{vaidyanathan2014htmlwidgets}. All code and many examples are available on
our github page: \url{https://github.com/krisrs1128/mvarVis}.

\subsection{Problem Motivation}

The analysis of complex data sets often requires a dimensionality reduction
step. Common methods for dimensionality reduction or feature extraction are
principal component analysis (PCA), canonical correlation analysis (CCA),
nonlinear multidimensional scaling (NMDS). While there are a number of R
packages for multivariate analysis, they often have a steep learning curve and
rarely share common code conventions. Further, data input and output formats
oftne vary largely from one package to the other. Finally, most packages
generate static results, an unnecessary limitation, in light of increasingly
accessible interactive visualization tools in the data analysis community.

mvarVis unifies different ways of performing multivariate analysis and
facilitates the process of generating visualizations. It is build on top of the
existing statistical packages, but adds a a layer of interactivity to the
resulting visualizations. With mvarVis, the user can easily add and display
supplemental information. Additionally, the package incorporates a bootstrapping
tool for resampling the input data. Bootstrapping is used to assess the
stability of the results of the analysis. The uncertainty of the dimensionality
reduction techniques can be estimated for example by inspecting the ordination
plots with bootstrapped samples.

\subsection{Principles}

The design of mvarVis reflects a few guiding principles,
\begin{itemize}
\item Identify ways to ease interpretation of existing implementations of
  multivariate analysis methods, but avoid developing new implementations or
  methods.
\item Leverage interactivity to facilitate users' navigation through alternative
  dimension reduction views.
\item Put checks in place to ensure correct conclusions are drawn from analysis
  results.
\end{itemize}

We detail these points below. The first point motivates links between mvarVis
and implementations in FactoMineR, ade4, and vegan. Our approach modularizes the
task into (1) converting the output from these packages to a consistent S4 class
and (2) creating visualization and interpretation tools that operate on this
class. By separating the implementation in this way, it is easy to extend
mvarVis on both fronts without risking incompatibilities elsewhere.

The second point was motivated by the observation that, in practice, to
understand multivariate analysis results, it is often necessary to iterate
across many types of supplemental information in order to characterize latent
dimensions. For example, when viewing the output of a PCA, it is common to color
points by many different variables which were not used in producing the
dimension reduction. By making this process interactive, we hope to help viewers
find meaningful patterns with less effort.

Finally, we take responsibility for correct interpretation of the displays
produced by mvarVis. One first check is to fix coordinate axis scalings to
reflect the proportion of variance explained by each axis
\citep{fukuyama2017multidomain}. Second, we provide methods for bootstrapping
results to assess ordination uncertainty -- points with wide contours should be
interpreted with caution \citep{efron1992bootstrap}.

\subsection{Implementation}

mvarVis uses an S4 class, which we call an mvarLayer, as a fundamental building
block unifying output from other ordination packages. Each layer is composed of
coordinates of projected points (a \texttt{coord} field) along with supplemental
information describing it (an \texttt{annotation} field). Different tables, or
types of coordinates, become different layers. For example, scores and loadings
in PCA are placed into separate layers. Or, in CCA, the two sets of scores
associated with each table might each be given a layer. By explicitly separating
types of points into layers, it becomes easy to customize attributes on a layer
by layer basis. This is a departure from traditional visualization packages --
for example, in ggplot2, the addition of a color scale to a plot modifies the
colors across all of the plot's components \citep{wickham2010ggplot2}.

This S4 class system also facilitates cross-talk across multivariate analysis
packages, by abstracting the mvarLayer information required for visualizations.
This makes it straightforwards to define a \texttt{convert\_to\_mvar()} utility
that converts the output from FactoMineR, ade4, and vegan into
\texttt{mvarTable}s. We have also provided a wrapper function,
\texttt{ordi\_wrapper()}, that wraps some of the most common functions in these
packages and then internally calls \texttt{convert\_to\_mvar()}. The currently
supported methods include,
\begin{itemize}
\item FactoMineR: Principal Components Analysis, Correspondence Analysis,
  Multiple Correspondence Analysis, Multiple Factor Anaysis, Dual Multiple
  Factor Anaysis, Hierarchical Multiple Factor Analysis, Factor Analysis of
  Mixed Data
\item ade4: Principal Components Analysis, Canonical Correlation Analysis,
  Procrustes Rotation, Correspondence Analysis, Multiple Correspondence
  Analysis, Fuzzy Correspondence Analysis, Nonsymmetric Correspondence Analysis,
  Fuzzy Principal Components Analysis, ``Hill-Smith'' Mixed-Type analysis
  \citep{hill1976principal}
\item vegan: Isomap, Detrended Correspondence Analysis, Constrained
  Correspondence Analysis, Principal Coordinates Analysis, Double Principle
  Coordinates Analysis, Nonmetric Multidimensional Scaling, Canonical
  Correlation Analysis
\end{itemize}

The interactive displays are created using D3, with communication between D3 and
R enabled by the htmlwidgets package \citep{vaidyanathan2014htmlwidgets}.
Hovering over points displays the raw data entry associated with them, an
instance of the same linking principle used in treelapse, and a control panel
allows customization of properties of the displayed points. In particular, the
color and size attributes can be made to reflect different fields in the
original data. Further, each point can be represented by either a circle, arrow,
or text label, since different types of plots often conventionally use different
displays -- for example, in \citep{dray2007ade4}, scores are usually given as
points, while loadings are given as arrows and labeled text.

Further, in many application it is of interest to assess the degree of stability
in the lower-dimensional representations of the data. The uncertainty in the
coordinates of the data points in the reduced space can be estimated by a
bootstrap \citep{daudin1988stability, chakerian2012computational}. Specifically,
in our implementation, we perform the following three steps,
\begin{enumerate}
\item Resample the input data matrix.
\item Apply a multivariate analysis method to each of the bootstrapped samples.
\item Plot bootstrapped points together in the reduced space.
\end{enumerate}
All steps are done automatically with a call to the
\texttt{bootstrap\_ordination()} function.

\subsection{Examples}

Example output from mvarVis is given in Figures \ref{fig:wine_interactive} and
\ref{fig:hmp_bootstrap}. Figure \ref{fig:wine_interaction} gives a screenshot of
an interactive display to an \texttt{mvarTable}, using the pedagogical wine data
set from \cite{le2008factominer}, which provides quantitative measurements from
wine tasters about drinks -- acidity and sweetness, for example -- from
different vineyards. The purpose of the data set is to identify characteristics
of the wines that vary substantially across vineyard regions or drink types, a
problem to which ordinary PCA is well suited. The loadings from this PCA is
given in Figure \ref{fig:wine_interactive}, the associated scores are available
on the mvarVis website.

\begin{figure}[ht]
  \centering
  \includegraphics[width=\textwidth]{figure/treelapse/wine_interactive}
  \caption{ An example of the \texttt{plot\_mvar\_d3()} function, using the wine
    dataset in FactoMineR. Hovering over points displays additional information
    about it, and alternative supplemental information can be displayed through
    the dropdown menu.
    \label{fig:wine_interactive}
  }
\end{figure}

Figure \ref{fig:hmp_bootstrap} gives example bootstrap ordination results on a
real microbiome data set. By applying this procedure, it becomes clear that the
positions of the skin samples are more variable than those for, say, the gut or
urogenital tract. This may be a result of differential sequence depths (or true
bacterial population levels) across different types of sites. Note that without
the bootstrapping step, this effect would not be clear.

\begin{figure}[ht]
  \centering
  \includegraphics[width=\textwidth]{figure/treelapse/hmp_bootstrap}
  \caption{
    Bootstrap Principal Coordinate Analysis with Morisita distance. Here we show
    an example of bootstrap PCoA using Morisita distance on the subset of Human
    Microbiome Project dataset \citep{human2012structure}. Outlined points
    represent the consensus positions across bootstrap replicates, and are close
    to one another then their microbiome abundance patterns are similar. The
    three panels give different ways of displaying the bootstrap samples which
    are provided by mvarVis: points, contours, and clouds.
    \label{fig:hmp_bootstrap}
  }
\end{figure}

\section{centroidview}

Centroidview is an R package for interactive visualization of hierarchical
clustering trees, available at \url{https://github.com/krisrs1128/centroidview}.
It is designed to facilitate inspection of subtree centroids, a useful follow-up
analysis of hierarchical clustering results, but which, in the absence of helper
utilities, can be complicated to implement and difficult to visually process.
That is, after clustering a collection of samples, it is often of interest to
characterize each cluster, rather than stopping at the fact that certain samples
were grouped together. The two most common approaches to this characterization
step are
\begin{itemize}
\item Centroids: We can use the average of quantitative features within
  different clusters as a kind of signature. Similarly, we can look at
  cross-tabulations of categorical features across clusters.
\item Heatmaps: By producing heatmaps of the raw data, with samples sorted
  according to leaf positions in the hierarchical clustering, structures between
  certain clusters and variables may become visibile.
\end{itemize}
Each approach has its advantages and disadvantages. Studying centroids can be
simpler than producing a full heatmap. It also avoids ambiguity of where
clusters begin and end. It can also be more informative in the case that
variables have a natural relation to one another, such as a time ordering. This
is because making visual comparisons of feature values as time series or
boxplots is more straightforward than differentiating color intensities across
heatmap cells. On the other hand, heatmaps avoid the issue of having to identify
the position at which to cut the hierarchical clustering tree. A related
advantage is that there are no problems caused by highly imbalanced cluster
sizes, which can complicate direct interpretation of centroids.

The main idea of centroidview is to blend these two approaches to inspecting
hierarchical clustering trees. Specifically, we apply the linking principle to a
combined view that includes the original hierarchical clustering tree, an
associated heatmap, and a centroid display. Hovering over a node in the tree
highlights the descendant samples in the heatmap. In the centroid display, a
line graph is produced whose $x$-axis indexes features and whose $y$-axis gives
the average value of the features among the samples descending from the
hovered-over node.

By interactively updating centroid summaries based on which nodes are hovered
over, centroidview allows the user to quickly evaluate the behavior of clusters
throughout the hierarchical clustering tree. Further, it becomes possible to
assess the sharpness of variation between neighboring nodes -- discontinuous
jumps are markers of clear clustering while smoother change in centroid shapes
suggests a gradient across samples.

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/treelapse/centroidveiw_linking}
  \caption{
    The most basic linking between the hierarchical clustering tree, a heatmap,
    and the centroids. Each row is one species, each column is one sample. The
    light regions in the heatmap correspond to drops in the time series. There
    are two other, undisplayed panels corresponding to the last two thirds of
    the heatmap columns.
    \label{fig:centroidview_linking}
  }
\end{figure}

Slight modifications to this display can enable more even sophisticated visual
comparisons. First, it is often of interest to compare two or more clusters to
one another. To this end, we allow the user to fix the view created by hovering
over one node, then layer on new views created by hovering over other nodes.
This view can clarify which features are responsible for differences between
subtrees.

A second modification allows a comparison of samples based on supplementary
features, which are not directly employed in the distanced used for clustering.
For example, suppose half of samples have property A, and the other half have
property B. It can be interesting to ask whether this property is associated
with features on which the clustering was defined. To this end, it is
interesting to tabulate the number of samples within each subcluster with each
property, looking for departures from independence. Visually, we can achieve
this using a bar plot that response to mouse movement along the tree. Each bar
is associated with one property, and the length of the bar is proportional to
the number of the hovered-over node's descendants with that property. See Figure
\ref{fig:centroidview_groups} for an example.

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/treelapse/centroidview_groups}
  \caption{
    An example of our approach to counting the numbers of subtree samples with
    different properties. Each sideways pair of bars corresponds to one
    ``property'' -- a taxonomic assignment of the species. Each color
    corresponds to one subtree of the hierarchical clustering. Evidenty, the
    currently highlighted green subtree consists of more Lachnospiraceae, and
    these are species which in one subject start at lower levels than those in
    the orange subtree, and disappear after the first antibiotic time course.
    \label{fig:centroidview_groups}
  }
\end{figure}


\subsection{\texttt{centroidview} case studies}

We apply this visualization approach to a hierarchical clustering on the
antibiotics data of \cite{dethlefsen2011incomplete}. This study investigated the
effects of antibiotics on the microbiome, especially as it relates to
instability and resilience. About 50 samples were collected longitudinally for
three subjects, and their trajectories of bacterial abundances were found using
16s sequencing.

Here we cluster species, using their abundances across individuals and
timepoints as features. We group species according to their taxonomic families.
Clustering returns collections of species with similar trajectories over time.
Our goal is to succinctly represent the essential trajectories that are visible
in the data, with the ultimate goal of interpreting these trajectories in
relation to theories of ecological dynamics.

An example display from centroid view applied to this data is given in Figure
\ref{fig:centroidview_differential}. Evidently, the orange and green subtrees
differentiate between species with differential responses to antibiotics in the
first and third subjects. In particular, for the first subject, the green
species are relatively unaffected by the second antibiotics, while the orange
species are. On the other hand, the same green species in the third subject seem
to recover to a lower level after the second antibiotics, while the orange
species gradually recover to their original level. Further, the bar plot
indicates that the the orange species consist of relatively more
Lachnospiraceae, while the green subtree mostly contains Ruminococcaceae.

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figure/treelapse/centroidview_differential}
  \caption{
    The complete centroidview display for the antibiotics data. Each row in the
    heatmap corresponds to one species, and each column is a sample. Samples are
    first grouped by person, then are sorted by time. The intensity of a cell in
    the heatmap reflects the abundance of that species in that sample. The
    hierarchical clustering tree is printed on the left. Subtree centroids are
    given in the panels along the top right, with one panel per subject and one
    line per subtree. Taxonomic breakdowns appear in the bottom right. Different
    colors distinguish different subtrees.
    \label{fig:centroidview_differential}
  }
\end{figure}

\section{Conclusion and Future Work}\label{conclusion}

Here, we have reviewed some fundamental principles of data visualization and
described their implementation in a new treelapse package. Further, we have
provided examples of the practical usefulness of these principles in real-world
data analysis situations.

This package has only developed basic ideas, and there are a number of
potentially useful extensions worth exploring. For example, we have not
considered the principle of arrangement in our visualizations
\citep{buja1996interactive}, though many of our conclusions were based on
comparing alternative selections of the same display. We could imagine faceting
our displays across groups to make these types of comparisons more accessible.
Further, we have only worked with the DOI distribution described in
\citep{heer2004doitrees}. It would be interesting to define a more statistical
notion of interest along nodes, based on cognostics, for example
\citep{hafen2013trelliscope, friedman2002john}. A simple extension could be to
allow graph layouts instead of trees in time and treebox displays, for data that
cannot be coerced into a hierarchical structure. Further, if these ideas turn
out to be useful in practice, it would be valuable to modularize the basic
visualization layouts and relationships into a library, allowing the wider
community to construct novel linked, interactive graphics with minimal effort.
Finally, formal quantitative assessments of interface design through a user
study could guide changes that improve the experience of practitioners.

In summary, we have built an easily accessible R package for visualization
techniques in a very specific methodology problem -- analysis of differential
abundance and dynamics in hierarchically structured data -- that appears in a
variety of application domains. We have leveraged a link between R and D3
\citep{vaidyanathan2014htmlwidgets} to create visualizations during the
exploratory phase of data analysis; in this way our work is a departure from the
culture of polished, journalistic visualizations prioritized by the D3 community
and is more closely aligned with the vision in \citep{de2003visual} of more
tightly integrating data visualization and statistical analysis techniques.
Finally, we have given a series of examples to demonstrate how the general
visualization techniques of focus-plus-context and linked brushing can be
practically incorporated into a range of practical analysis workflows, from
studying the impact of bacteria on human health to better allocating units in
commuter bikesharing systems.
