\documentclass{report}
\usepackage{suthesis-2e}
\usepackage{amsmath, amssymb, amsfonts}
\usepackage{natbib}
\usepackage{graphicx}
\usepackage{hyperref}
\input{preamble.tex}

\title{Interactive Visualization and Probabilistic Models \\
  for the Microbiome}
\author{Kris Sankaran}
\dept{Statistics}

\begin{document}
\maketitle
\tableofcontents

\chapter{Multitable Methods and Microbiome Data Integration}

\section{Classical multivariate statistics}

\subsection{PCA}

\subsection{CCA}
\label{sec:cca}

\subsection{Co-Inertia Analysis}

\subsection{MFA}
\label{sec:mfa}

\subsection{PCA-IV}

\subsection{Partial Triadic Analysis}

\subsection{Statico and Costatis}

\subsection{Reduced-rank regression}
\label{sec:rr-reg}

\section{Modern multivariate methods}

\subsection{PLS}
\label{sec:PLS}

\subsection{CCpnA}
\label{sec:canonical-correspondence}

\subsection{Curds \& Whey (C\&W)}
\label{sec:cw}

\section{Methods from machine learning}

\subsection{Kernel Canonical Correlation Analysis (KCCA)}

\subsection{Penalized Matrix Decomposition}

\subsection{Covariate-assisted Spectral Clustering}

\subsection{Multitask learning}
\label{sec:multitask}

\subsubsection{Bayesian multitask learning}

\subsubsection{$\ell^{2, 1}$ Multitask learning}

\subsubsection{Graph-based Multitask Learning}

\subsection{Matrix Factorization}

\chapter{Interactive Visualization of Microbial Counts}

We introduce methods for visualization of data structured along trees,
especially hierarchically structured collections of time series. We hope both to
characterize generically useful techniques for interactively visualizing
hierarchical data and to offer practical tools for implementing such displays.
To this end, we identify questions that often emerge when working with
hierarchical data and provide an R package to simplify their investigation
\citep{ihaka1996r}.

In particular, we adapt the visualization principles of focus-plus-context and
linking to the study of tree-structured data \citep{buja1996interactive,
  becker1987brushing}. Our motivating application is to the analysis of
bacterial time series, where an evolutionary tree relating bacteria is available
a priori. However, we have identified common problem types where, if a tree is
not directly available, it can be constructed from data and then studied using
our techniques.

We have implemented our visualizations in D3, but encapsulated in an R package,
called treelapse, to facilitate rapid turnover from data preparation and
modeling to interactive exploration, and vice versa. The package source code and
scripts for various case studies, including all those presented below, are
available at
\href{http://krisrs1128.github.io/treelapse}{http://krisrs1128.github.io/treelapse}
We hope this package encourages data analysts to work at the border between data
modeling and visualization, and more generally empowers a wider audience to
apply less widely known, but powerful, visualization ideas.

In summary, our key contributions are,
\begin{itemize}
\item Proposals for visualizing hierarchically structured data, based on
  principles from the data visualization community.
\item The implementation of these proposals in a publicly available R package.
\item The illustration of the wide reach of hierarchical data visualization,
  through case studies in both scientific and societal contexts.
\end{itemize}

The chapter is organized as follows. First, we describe our motivating application
to the microbiome and the associated generic analysis tasks. Next, we review the
underlying visualization principles behind our contributions. Then we then
connect these principles to analysis tasks we identified earlier, describing in
detail the visualization methods we have implemented in treelapse. We close with
several case studies using publicly available data across both microbiome and
non-microbiome related applications.

\section{Problem Motivation}

The two essential microbiome analysis problems that motivated our work are the
tree-structured differential abundance and differential dynamics problems. In
the differential abundance problem, we attempt to compare the abundances of
individual bacteria across experimental conditions -- for example, treatment vs.
control or healthy vs. diseased. This is the microbiome analog of differential
expression analysis in genomics \citep{anders2010differential}. We prepend the
description ``tree-structured'' because, in practice, researchers generate
interpretations about intermediate taxonomic orders -- it is more interesting to
discover novel behavior taxonomic levels between high-order phyla and low-level
species. Hence, we frame the tree-structured differential abundance problem as
the question of identifying the largest taxonomic subtree whose associated
bacteria are differentially abundant.

In the tree-structured bacterial dynamics problem, the goal is to describe
changes in bacterial abundances in an environment over time. As in the
differential abundance problem, it is useful if these descriptions can be given
at the highest subtree at which the pattern appears. Specific questions of
interest often have an ecological flavor. For example, researchers are often
interested in understanding how bacterial populations respond to sudden or
gradual environmental changes or how species fill, drop out from, or compete for
environmental niches. Medically, these questions are important for illuminating
the impact of antibiotic time courses or diet changes, for example.

To unify the tree-structured differential abundance and bacterial dynamics
problems, we identify the data with a collection of random variables indexed by
nodes in a prespecified tree structure. In the differential abundance problem,
each random variable lives in $\mathbb{R}^{G}$ where $G$ is the number of
groups being compared. Each coordinate represents the abundance for that group,
and a node exhibits differential abundance when the coordinates are drawn from
different distributions. On the other hand, in the bacterial dynamics problem,
each random variable is a time series, living in $\mathbb{R}^{T}$.

\section{Visualization principles}

Now that we have specified the essential questions of interest, we survey some
ideas from the visualization literature that can be applied to answer them. As
the core difficulty is high-dimensionality, it should be no surprise that the
techniques we adapt come from the literature on high-dimensional data
visualization, which has enjoyed rapid progress in the last 25 years. Modern
research in this domain develops abstractions and taxonomies for guiding
visualization designs so that they most effectively communicates properties of
the data to their intended audience. A major push in this body of work explores
the potential for interactivity to improve many stages of the data analysis
process, from preliminary data preparation, to refinement and navigation across
views, to final sharing and annotation of results \citep{heer2012taxonomy}.
Further work has attempted to bridge the gap between statistical analysis and
data visualization methodology, both of which provide techniques for learning
from high-dimensional data \citep{de2003visual}.

\subsection{Focus plus context}

The problem structures most relevant to our study are tree and temporal
structure, and the visualization community has various ways of reasoning about
these data, see \citep{graham2010survey, aigner2011visualization} for detailed
surveys. From this literature, our approach is most directly informed by the
focus-plus context and linking principles, which we briefly review here. The
focus-plus-context principle is that large collections of visual elements can be
studied at multiple scales, by simultaneously focusing" on a few elements of
interest and maintaining the ``context'' of the coarser-scale background. A
simple example of this idea is to include a search box that highlights matching
samples (focus) and mutes the rest (context). Two more sophisticated methods
anchored in this idea are timeboxes and Degree-of-Interest (DOI) trees; both are
central to the proposals in treelapse \citep{hochheiser2004dynamic,
  heer2004doitrees}. In timeboxes, a collection of time series are graphically
queried using interactive brushes. Series that pass through all of the
user-specified brushes are highlighted, and the rest are faded to the
background. Hence, time series meeting the constraints imposed by the brushes
are focused, while the remainder are de-emphasized, though they remain present
as context. This method can be interpreted programmatically as the visual analog
of a database query, or probabilistically as the conditional distribution for
the full series, given it passes through certain bounds.

In DOI trees, the viewer's attention is focused on a collection of high-interest
nodes, while the remaining lower-interest nodes are left on the fringes as
context. The implementation is modularized into two tasks -- the determination
of a DOI distribution over nodes in the tree and visual layout of a tree given
DOI assignments. The DOI distribution used in \citep{heer2004doitrees} places
maximal interest on the node that the user had most recently clicked, along with
all ancestors. The DOI for all other nodes is defined as the distance to the
closest maximal interest node. The layout step then trims low-interest subtrees
until the remaining nodes fit within a given screen size. By adjusting the
minimal DOI below which nodes are hidden, the user can transition between
node-specific and full-tree scales.

\subsection{Linking}

In linking, alternative representations of the same samples are placed
side-by-side in order to display covariation across views. A canonical
application is to linked scatterplot brushing \citep{becker1987brushing}. Here,
a scatterplot matrix gives the relationship between all pairs of variables.
~Points brushed in one scatterplot are then highlighted in all others. For
example, this helps the user determine whether an outlier in one dimension is an
outlier in others. Another instance of this idea links the results of
dimensionality reduction methods to displays of the raw data, as implemented by
XGobi and Cranvas, for example \citep{xie2013cranvas, swayne1998xgobi}. As in
timeboxes, linking can be interpreted as database queries or conditional
probabilities: given a subset of the series after conditioning on the values for
one set of features, what are the values for a second set
\citep{buja1996interactive}?

\section{Design of the \texttt{treelapse} package}

Our first proposed visualization technique is a minor modification of the DOI
tree. The standard DOI tree definition does not have any notion of data defined
at nodes, it is only used a device for navigating tree structures. A trivial
extension can encode scalar data at nodes: have the node radius reflect the
associated scalar value. To reinforce this effect, we can adjust the width of
the parent edge. When parent nodes have values equal to the sum of their
children, this creates the effect of values ``flowing'' from the root to leaves.
To help viewers make use of their domain knowledge, we have included a search
box that highlights paths to nodes with matching terms. Edges are ordered from
widest on the left to narrowest on the right. While this method can only
represent a single scalar-value per tree node, it suggests an approach to the
tree-structured differential abundance problem, which we call the DOI sankey.

In the DOI sankey, we split each edge in the DOI Tree across different groups.
For example, suppose we have the average counts for treatment and control groups
at each tree node. Every edge in the tree is split into two colors\footnote{We
  use the colorbrewer palette to facilitate readability
  \citep{brewer2003colorbrewer}}, with relative widths of the different colors
reflecting differences in sizes for the two groups. The overall width of each
edge represents the sum of values across all groups.

This display is designed to facilitate investigation of the tree structured
differential abundance question. For example, for a single node and a single
group, first compute the average abundance at that node among all samples in
that group. This will give the width for that group's color on the edge leading
to the specified node. Differentially abundant subtrees then correspond to
subtrees where some colors occupy more space than others. That is, this
representation makes it easier to identify points where the ``flows'' for
different groups diverge -- the colors begin to separate. The DOI principle
assists navigation across the tree structure, allowing focus on individual flow
structures without losing broader tree context.

Our third display is directed at the bacterial dynamics question. Here, two
panels are arranged one over the other; one displays all time series, while the
other displays all tree nodes, with node sizes reflecting the value at that node
averaged across all time points. In the time series panel, we have directly
implemented the timeboxes idea. We then link the panels: when a set of series is
highlighted by the timeboxes, the associated tree nodes are also highlighted.
For example, timeboxes can be used to focus on a set of series with specific
shape -- increased abundance after an ecological shock, for example -- and
identify along what subtrees this pattern is present. To further focus on
specific elements, a pan-zoom scented widget is provided
\citep{willett2007scented}. The widget is a miniature version of the full time
series panel, equipped with a single brush whose extent specifies the limits in
the main time series panel. As in the DOI trees and sankeys, a search bar can be
used to highlight those series of interest a priori.

The final display currently implemented in the package is the natural converse
of the timebox trees display. Rather than defining visual queries in terms of
time series, it defines queries using nodes in the tree. Rather than focusing on
the intersection of brushes, as in timebox trees, we focus on the union of
brushed over nodes. This allows us to highlight series associated with nodes on
distant subtrees. This display is also suited for the bacterial dynamics
problem. For example, by highlighting all nodes at one taxonomic level in the
tree, we can easily summarize the time series pattern for all the taxa at that
level. Alternatively, focusing on all the children below a single node makes it
possible to see how much correlation and competition there is between
taxonomically similar bacteria.

To be practically useful, the resulting visualizations must respond fluidly to
user interaction. As the data increase in scale, this fluidity can deteriorate
for two reasons. First, rendering many SVG elements in a framework like D3 is
costly\citep{johnson2008scalability}. While it is possible to use alternatives
-- HTML5's Canvas, for example -- it is often more challenging to implement
complex interactive behavior through them. Second, some of the dynamic queries
require a search over a many elements. These limitations are most pronounced in
the timebox tree display, which must search through all timepoints among all
time series whenever the brush is moved. The first, but not the second, concern
applies to treeboxes, while neither applies to DOI trees. Nonetheless, we feel
comfortable recommending timebox trees for data on the order of 500 tree tips
and 50 timepoints.


\section{\texttt{treelapse} case studies}

We now delve into applications on real data. Our goals are to illustrate
potential workflows that incorporate treelapse, describe the formulation of
questions that can be naturally investigated with our methods, and provide
example interpretations on treelapse output. Our examples are also chosen to
reflect the range of problem domains to which the package can be applied --
though it was motivated by applications to the microbiome, it is not tied to it.
More importantly, we argue that the visualization principles reviewed above can
substantively improve the practice of data analysis in the class of problems to
which we have limited ourselves.

\begin{table}
\centering
\begin{tabular}{|l|l|l|}
  \hline
  Data            & Number of timepoints & Number of nodes \\
  \hline
  Antibiotics     & 56                 & 386             \\
  Preterm births  & 216                & 318             \\
  Housing prices  & 254                & 944             \\
  Bikesharing     & 24                 & 819             \\
  Global patterns & 500                & 51             \\
  \hline
\end{tabular}
\caption{Problem dimensions for each of the case studies. For problems with
  dimensions larger than that in the housing prices problem, we recommend an
  initial summarization or filtering step to prevent performance
  issues. \label{problem-scaling}}
\end{table}

\subsection{Bacterial dynamics in antibiotics time courses}
\label{bacterial-dynamics-of-antibiotics-time-courses}

\citet{dethlefsen2008pervasive} investigated the effect of antibiotics on
bacterial community composition from an ecological perspective. The study tracks
the microbiome of three patients across ten months, with two five-day antibiotic
time courses separated by 6 months. Discerning the variation in resilience
across bacteria is important, considering the role of bacteria in health and
not just disease.

We approach the data using the linked time and treebox views, after first
filtering low variance taxa and taking an \(\text{asinh}\) transformation. An
initial view, Figure \ref{fig:antibioticoverview}, reveals two dramatic drops in
the overall bacterial abundance time series during the antibiotics time courses.
Two more subtle effects are also suggested from this view,

\begin{itemize}
\item
  The second antibiotic treatment seems to have a more lasting effect,
  as the series take longer to return to their original values.
\item
  Some high level taxa appear relatively unaffected by the first
  antibiotic treatment. By more closely inspecting the display, we are
  able to identify these as members of the Bacteroidetes phylum, see
  Figure \ref{fig:antibioticbacteroidetes}.
\end{itemize}

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/annotated_antibiotic_overview}

}

\caption{Here we display the primary timebox tree view of the antibiotics data
  set from \citep{dethlefsen2008pervasive}, annotated with the main components of the
  visualization. The tree at the top is a taxonomic tree of all the bacteria
  contained within the sample, and it is visually linked to the time series at
  the bottom: each node in the tree corresponds to a path among the time series.
  The selection brush is used to focus attention on the time series that go
  through it -- these are highlighted in blue -- and other brushes can be added
  using a button not displayed here. The pan-zoom widget at the top right is
  used to update the scales of the main time series display so that only
  particular time windows and $y$-axis ranges are visible. This view is the
  basis for all the timebox tree and treebox displays that appear
  below. \label{fig:antibioticoverview}}
\end{figure}

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/antibiotic_bacteroidetes}

}

\caption{Introducing a second box into the timebox display identifies the
  Bacteroidetes as a taxon only mildly impacted by antibiotics. The layout is
  identical to Figure 1, except two small brushes are placed over the time
  series between 10 and 20 days, and now only those time series and
  corresponding nodes in the tree are highlighted in blue. Further, the user has
  hovered over the top blue node in the tree, revealing the taxonomic identity
  of these series. Hence, brushing the time series and linking with the tree can
  be used to discover and characterize notable variation within the
  data.\label{fig:antibioticbacteroidetes}}
\end{figure}

Next, using the scented widget, we focus on the window around the second
antibiotic treatment. We apply the treebox display to compare then behavior of
different families of Firmicutes, Lachnospiraceae and Ruminoccocus. We suspect
that these taxa are associated with the delayed recovery after the second time
course. To investigate this, we input these family names in the search box to
isolate their positions on the tree; then we apply brushes to highlight the
series that contribute to these higher-level families. The resulting view is
given by Figure \ref{fig:antibioticfirmicutes}

\begin{figure}
{\centering \includegraphics[width=375px]{figure/treelapse/antibiotic_firmicutes}}
\caption{Zooming into the second antibiotic time course and highlighting the
  Lachnospiraceae and Ruminococcus, we see that the Ruminoccocus took more time
  to recover to pre-treatment levels. Here, the red lines and nodes are those
  that match the text search provided by the user in a search box just below the
  figure (not displayed here). Hovering the mouse over these lines provides
  their identities -- the top red line is Lachnospiraceae, and the bottom red
  line is Ruminoccocus. Note that the brush in the treebox display is located
  over the tree, rather than over the time series. In particular, the search box
  and interactive brushing can be combined to interrogate hypotheses of a priori
  interest. \label{fig:antibioticfirmicutes}}
\end{figure}

Alternatively, we can summarize each node by the average across its descendants
-- this brings attention to individual bacteria that may be underlying some of
the broader taxonomic patterns we have noted when studying the subtree sums. For
example, in Figure \ref{fig:ruminococcus}, we highlight all families below order
Ruminococcus, suggesting that the decrease due to antibiotics occurs uniformly
across almost all families. A point that was not evident in the earlier
sum-across-descendants view is that, after the second treatment of antibiotics,
a few of the Ruminoccocus families recover more rapidly than the rest, for
example the Unc095d3 (highlighted in red) are only briefly affected. In
contrast, most families seem to recover in unison after the first treatment.

\begin{figure}

{\centering \includegraphics[width=375px]{figure/treelapse/ruminococcus}}
\caption{By hovering over the Ruminoccocus branches, we see that there is a
  prolonged effect of the antibiotics time courses more or less uniformly across
  the lower taxonomic orders. The graphical elements are the same as before,
  except the user has searched for Ruminoccocus and species Unc095d3, which has
  the highest average abundance within this taxon. By displaying averages rather
  than sums, we see that the effect of antibiotics visible at higher taxonomic
  orders is not created by a single dominant species becoming less abundant, but
  rather the decline in populations across all descendant species. The same
  display applied to different data can yield different
  insights.}\label{fig:ruminococcus}
\end{figure}

Further, note that in this subtree averages view, the tree display has changed.
This is because, at each branching point, we place the node with larger average
value on the left. Figure \ref{fig:verrucomicrobiae} notes that the nodes at
the far left in the tree are associated with phylum Verrucomicrobiae,
corresponding to a large average abundance across time points. This phylum had
been previously obfuscated -- because there are not many leaves associated with
this phylum, the sum was small. The abundance of these bacteria seems to
\emph{increase} after the first antibiotics treatment. Be cautious, however,
that the average over only a few Verrucomicrobiae species will be a more
variable estimate than the averages over the more prevalent phyla.

\begin{figure}
{\centering \includegraphics[width=375px]{figure/treelapse/verrucomicrobiae}}
\caption{The subtree averages aggregation brings attention to the
  Verrucomicrobiae, which though only present as a few species, are each rather
  abundant. In particular, they seem to increase after the first antibiotic time
  course, which occurs between days 15 and 20. This view was generated by
  placing a brush over the branch on the far left, which has those nodes with
  the largest averages across all timepoints. The user's mouse is over the blue
  series, which brings up the associated taxonomic label. The determination of
  species whose abundances increase during antibiotics, which would require many
  hypothesis tests using a more standard approach, becomes quickly apparent
  via interactive visualization.}\label{fig:verrucomicrobiae}
\end{figure}

\subsection{Differential Bacterial Abundance and Preterm Births}
\label{differential-bacterial-abundance-and-preterm-births}

\citet{digiulio2015temporal}
tracked the abundance of bacteria in the vaginal microbiome during
pregnancy in an effort to study relationships between bacterial
community composition and preterm birth. Ideally, it would be possible
to develop clear bacterial signatures associated with preterm births.

Unlike the antibiotics study, we have measurements across more
individuals than we could reasonably inspect one at a time. While we
could average across all individuals, we will take our cue from
\citep{digiulio2015temporal}
place each sample into one of five Community State Types (CSTs),
identified via k-medoids. In that study, a linear model identified one
of these CSTs (CST 4) as significantly more diverse, further it appeared
associated with preterm births. Here, we corroborate this finding using
exploratory views.

Therefore, our focus here is on the differential abundance question,
rather than dynamics. We would like to provide visual representations of
differential abundance across CSTs and also between preterm and non-preterm
births. \citet{digiulio2015temporal} interpreted the CSTs using a heatmap, with
bacteria ordered according to a hierarchical clustering. By using the DOI sankey
instead, we can interpret the CSTs in their taxonomic context and at multiple
scales of taxonomic resolution. Further, while
\citet{digiulio2015temporal} focused on identifying associations between
preterm births and CSTs -- presumably because testing individual bacteria loses
power -- we can compare bacterial abundances between preterm and non-preterm
samples along subtrees, without requiring CSTs as an intermediary.

In Figure \ref{fig:pretermcsts}, we compare the 5 CSTs according to
their values along the subtree. Specifically, we took the average of all
samples within each CST to define values at the (species-level) leaf
nodes, and then aggregated the averages up to the root. It is
immediately clear that samples from CST 4 have much more taxonomic
diversity. Further, focusing on the Lactobacillaceae family, we note
that the differential abundance of these bacteria distinguishes the
remaining CSTs, see Figure \ref{fig:pretermcstslacto}.

Alternatively, in Figure \ref{fig:pretermpreterm}, we avoid working with CSTs,
displaying instead averages among samples associated with either preterm or term
births. The green and yellow edges are associated with preterm births -- we see
that they contribute more weight to phyla outside the Firmicutes. This is
consistent with the claim that CST 4, the most diverse of the CSTs, is
associated with preterm births.

\begin{figure}
{\centering \includegraphics[width=375px]{figure/treelapse/preterm_csts}}
\caption{The increased diversity among samples in community state type (CST) 4
  is represented by the relatively larger contribution of red edges to branches
  outside of the Firmicutes. This display shows the top of the DOI sankey
  visualization of the preterm birth data studied in
  \citep{digiulio2015temporal}. The root of the tree is the taxonomic kingdom
  Bacteria, its children are labeled according to their phylum names. Each color
  within a branch is associated with CST, and the width of the associated color
  corresponds to the average abundance of that taxonomic group among all samples
  belonging to the corresponding CST, as indicated by the legend at the left.
  Phyla are sorted from most to least abundant. This initial display of the DOI
  sankey provides a summary of overall abundances across taxonomic groups and
  CSTs, and suggests subtrees to navigate into, to extract more detailed
  abundance information.
}\label{fig:pretermcsts}
\end{figure}

\begin{figure}
{\centering \includegraphics[width=375px]{figure/treelapse/preterm_csts_lacto}}
\caption{Zooming into the Lactobacillaceae family, we notice that the difference
  between the remaining four CSTs is related to which types of Lactobacillus are
  most prominent. The DOI sankey refocuses the tree around the last group that
  was clicked on, in this case showing more detail about order Lactobacillales
  and its descendants. The overview display can be recovered by navigating back
  up to higher-level taxa. Hence, it is possible to navigate between broad
  overview and detailed displays in a way that facilitates interpretation of
  results from statistical analysis.}\label{fig:pretermcstslacto}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=300px]{figure/treelapse/preterm_preterm}
\caption{Samples with high levels of phyla other than Firmicutes appear to be
  related to preterm births. We again display the preterm birth data with
  a DOI sankey, but instead of grouping samples according to statistically
  generated CSTs, we directly assign samples to preterm vs. not preterm
  according to whether the mother eventually had a preterm birth. These new
  states are visible in the updated legend. Through interactivity, it becomes
  possible to develop meaningful visual summaries even before calculating formal
  statistical ones.}
\label{fig:pretermpreterm}
\end{figure}

\subsection{Sources of Variation in Bikesharing
Demand}\label{bikesharing-study}

Our next example is a study in bikesharing demand, included as an example of
analyzing collections of time series when there is no obvious hierarchical
structure a priori. The data are available at the UCI Repository and were
originally collected by a Washington D.C.-based bikesharing system for use in a
Kaggle prediction competition. The data are hourly measurements of bike demand,
aggregated across all bikesharing stations, over two years, along with
supplemental weather data. In the competition, participants were asked to
predict the hourly demand on a held-out test set. Here, we adopt a descriptive
view instead, attempting to characterize factors associated with variation in
bikesharing demand.

We study this problem as one of describing a large collection of related time
series. Here, we consider the demand during a single day to be one time series;
this is a natural choice considering the daily periodicity of bike demand. To
arrange these daily series along an interpretable tree structure, we apply a
regression tree relating the supplemental data to the bikesharing demand
\citep{breiman1984classification}. In more detail, we built this tree by noting
the ``two table'' structure of this problem: one describes bike demand, the
other holds the supplemental data. In both, the rows index days, while the
columns index either hours or supplemental features. Our tree is the trained
regression tree after predicting demand at 8AM based on the supplemental data.
We choose this response because (1) we need a univariate response in order to
apply regression trees and (2) the more straightforwards reduction to
daily-average-demand fails to distinguish between weekdays and weekends, whose
series appear qualitatively very different from each other.

Given this response, the first split in the regression tree is (unsurprisingly)
the difference between weekends and weekdays. This is emphasized in Figures
\ref{fig:working} and \ref{fig:weekend}, respectively; using timeboxes to
isolate the two types of series highlight the left and right sides of the tree,
respectively. For a more subtle effect, we select the internal nodes associated
with the first split below the weekday vs. weekend split; these are given in
Figures \ref{fig:weekday2011} and \ref{fig:weekday2012}. This suggests that
weekday demand increased during the second year.

\begin{figure}
{\centering \includegraphics[width=375px]{figure/treelapse/working}}
\caption{The two peaks at rush hour distinguish weekday series from the rest.
  through the timebox tree view. The display is the same type of timebox tree
  view introduced in Figure \ref{fig:antibioticoverview}, but applied to the
  bikesharing data, where the time axis represents the time of day and the
  $y$-axis gives bikesharing demand. Each series is the bikesharing demand for a
  single day, over the course of two years. The tree now corresponds to the
  regression tree generated by predicting demand at 8am using supplementary
  data. Two brushes are introduced to highlight the double peaks corresponding
  to rush hours on weekdays. We see that although hierarchical structure was not
  present immediately in the bikesharing data, it is useful to introduce and
  interpret such structure by combining regression and visualization
  methodology.
}\label{fig:working}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=375px]{figure/treelapse/weekend}
    \caption{By adjusting the two brushes in Figure \ref{fig:working}, we see that
      unlike weekday demand, weekend demand is unimodal. The few weekday series with
      unimodal series seem to be associated with holidays. This is the case for New
      Years' Eve, which is currently hovered over in the tree. The ease of
      transitioning from Figure \ref{fig:working} to this display indicates the
      importances of brushing in interaction.}\label{fig:weekend}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=375px]{figure/treelapse/weekday_2011}
    \caption{Weekday demand appears larger in 2012 than 2011 -- compare with Figure 
      \ref{fig:weekday2012}. Here, a brushes are introduced over the tree to see the
      series associated with a particular split point. The red nodes are the results
      of searches over the two nodes that are children of this split point. The fact
      that the \texttt{yr>=0.5} selected line is larger than the \texttt{yr<0.5}
      line means that demand was larger in 2012. In combination with searching and
      treeboxes, it is possible to interpret more subtle split points in the decision
      tree.} \label{fig:weekday2011}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=375px]{figure/treelapse/weekday_2012}
  \caption{Weekday demand increased in 2012 -- compare with 
    Figure \ref{fig:weekday2011}. The search terms are the same as in that figure,
    but the subtree associated with the 2012 split point is highlighted, using the
    union of two boxes. Note that unlike timebox trees, which highlighted series
    lying through the intersection of brushes, treeboxes highlight series within the
    union of brushes.}\label{fig:weekday2012}
\end{figure}

In contrast to these general questions about daily demand, we could ask
a more granular question about specific time windows. For example, what
characterizes days on which there is larger than average demand after
midnight? We can select these series after first zooming into this time window.
Figure \ref{fig:warmweekend} reveals that the highlighted series are associated
with the warm-weekend split, which seems quite reasonable in retrospect.

\begin{figure}
{\centering \includegraphics[width=375px]{figure/treelapse/warm_weekend}}
\caption{The samples with highest night demand tend to fall on warm
  weekends. Here, the pan-zoom widget has been used to adjust both time and
  demand axes to narrow on a specific window of interest. Brushes are drawn over
  the larger among these series, and the corresponding tree nodes are located
  close to one another, in the part of the tree corresponding to the warm / cold
  average temperature split. More generally, panning and zooming allows
  navigation between focus and context.}\label{fig:warmweekend}
\end{figure}

Finally, we can study the behavior of the regression tree itself using
the DOI sankey. Here, we group samples according to their quintile of
8AM demand and then count the abundance of the groups flowing down
different branches. We find that the quintiles are each rather strongly
separated after descending even a few steps down the regression tree --
for example, Figures \ref{fig:weekday2011} and \ref{fig:weekday2012}
focus on 2011 vs. 2012 split among weekday samples, showing that this
split distinguishes between samples falling in the second and third
quintiles of 8AM demand.

\begin{figure}
{\centering \includegraphics[width=375px]{figure/treelapse/bike_sankey}}
\caption{So far, we have focused on the timebox tree and treebox representations
  of the bikesharing data -- a complementary view is provided by the DOI sankey.
  Here, the tree is the result of the regression tree procedure, while the
  colors represent particular quantiles of 8am demand. This allows the
  determination of which split descendants are associated with low or high
  demand.}\label{fig:bikesankey}
\end{figure}

This interactive representation of regression trees is potentially more
useful on larger trees that cannot be easily parsed in a single view; in
this sense the bikesharing tree is relatively simple. In our ideal data
analysis workflow, we imagine the analyst applying interactive
visualization and modeling techniques in an iterative, nonlinear
fashion, in the spirit of \citep{de2003visual}.

\section{Design of the \texttt{hclustvis} package}

Reading a panel
- heatmap, tree, centroids
- time series -> faceting
- breakdown across taxonomies in bottom right
- compare clusters side by side

\section{\texttt{hclustvis} case studies}

\chapter{Mixed-membership Models for Microbiome Time Courses}

\section{Problem description}

\section{Available methods}

\subsection{Latent Dirichlet Allocation}

\subsection{Dynamic unigram and Topic Models}

\subsection{Nonnegative Matrix Factorization}

\section{Simulations}

\section{Case study}

\chapter{Inference of Dynamic Regimes in the Microbiome}

\section{Problem description}

\section{Methods baseline}

\subsection{Hierarchical Clustering}

- Limitations, advantages
- Interpretation of centroids
- Choices: transformation, distances

- Review of hierarchical clustering

- Distances
Using euclidean distance
Using innovations
Using jaccard distance
Combining distances

General question: How to choose between these distances?

Interpretation
- ordinary averages within components
- proportions nonzero

\subsection{Recursive partitioning}

- Variation on hierarchical clustering, to provide temporal structure
- Brief description of recursive partitioning
- Application to the antibiotics data: provide heatmap and interpretation, at
several levels of complexity.

\section{Temporal probabilistic models}

\subsection{Linear dynamical systems}

- Review LDS graphical model
- Derivation of Kalman Filter / Smoother
  + Kalman filtering pseudocode
  + Kalman smoothing pseudocode
  + Link to some actual code

\subsection{Gaussian processes}
- Review GP graphical model
- Basic distinctions from previous models
- Description of GP posterior
  + GP pseudocode
  + Link to some actual code

\section{Temporal mixture models}

\subsection{Hidden Markov Modeling}

- Probabilistic approach that incorporates time structure directly
- Draw the graphical model

\subsubsection{Standard HMMs}

- Write the model and provide an interpretation
- Application to antibiotics: heatmap, interpretation, and transition
probability matrix
- In order to understand extensions, briefly review actual inferential mechanism
- What kinds of limitations have we encountered?

\subsubsection{Sticky HMMs}

- Modified model
- Simulation example
- Antibiotics case study: Transition probabilities and heatmap. Interpretation
of transition probabilities. Contrast with hierarchical clsutering and recursive
partitioning
- Review inferential procedure (block sampling)
  + Link to code
  + Algorithm psueodocode
  + Derivation of some nonobvious steps

\subsubsection{Sticky HDP-HMMs}

- Write the new model
- Simulation example
- Antibiotics case study
- Review inferential procedures (direct assignment and block sampling)
  + Algorithm psueodocode
  + Derivation of some nonobvious steps
  + Link to code

\subsection{Mixture of Experts}

- Switching state space models
   + Model description
   + Simulation example
   + Description of variational inference
- Mixture of Gaussian processes
  + Model description
  + Simulation examples
  + Description of variational inference

\section{Alternative probabilistic models}

\subsection{Changepoint detection}

- Description of the model
- Summary of the empirical bayes idea
- Summary of dynamic programming, and provide row and column updates
  + Maybe provide actual pseudocode?
- Would be nice to include a simulation / application to antibiotics

\subsection{Accounting for zero inflation}

- The dynamic tobit model
  + Brief summary of Glasby Nevison
  + Explanation and simulation using the scan sampler

\chapter{Conclusion}

\bibliographystyle{plainnat}
\bibliography{refs.bib}

\end{document}
